// fn-field-message/fesm2022/fn-field-message.mjs
import { CommonModule } from '@angular/common';
import * as i0 from '@angular/core';
import { inject, Input, Component } from '@angular/core';
import { TranslateService, TranslateModule } from '@ngx-translate/core';
import { FNLabel } from 'fn-label';
import { Subscription, merge } from 'rxjs';
import { startWith } from 'rxjs/operators';

class FNFieldMessage {
    cdr;
    field;
    control;
    hasFocus = false;
    translate = inject(TranslateService);
    subs = new Subscription();
    constructor(cdr) {
        this.cdr = cdr;
    }
    ngOnInit() {
        if (!this.control) {
            // ensure at least initial render works
            this.cdr.detectChanges();
            return;
        }
        // startWith ensures initial evaluation immediately
        const val$ = this.control.valueChanges.pipe(startWith(this.control.value));
        const status$ = this.control.statusChanges.pipe(startWith(this.control.status));
        // merge both streams so any change forces update
        this.subs.add(merge(val$, status$).subscribe(() => {
            // detectChanges to ensure inner OnPush children get updated immediately
            try {
                this.cdr.detectChanges();
            }
            catch {
                // ignore change-detection errors if they happen during CD cycle
                this.cdr.markForCheck();
            }
        }));
    }
    ngOnChanges(changes) {
        if (changes['hasFocus'] || changes['control']) {
            // ensure re-evaluation on parent-driven changes
            try {
                this.cdr.detectChanges();
            }
            catch {
                this.cdr.markForCheck();
            }
        }
    }
    ngOnDestroy() {
        this.subs.unsubscribe();
    }
    getFieldMessage() {
        // defensive: no control => show helper (initial server-side or static)
        if (!this.control)
            return this.field?.helperText || '';
        // show validation error message when touched + invalid
        if (this.control.touched && this.control.errors) {
            for (const key of Object.keys(this.control.errors)) {
                // Check if the error object has a 'message' property (for dynamic error messages)
                const errorValue = this.control.errors[key];
                if (errorValue && typeof errorValue === 'object' && errorValue.message) {
                    return errorValue.message;
                }
                // Fallback to field.errors[key] for static error messages
                if (this.field?.errors?.[key]) {
                    return this.field.errors[key];
                }
            }
            // fallback error message if no mapping provided
            return (this.field?.errors?.default || 'Please enter ' + this.translate.instant(this.field?.label));
        }
        // default (untouched or other states): helper text
        return this.field?.helperText || '';
    }
    get statusColor() {
        if (this.control?.disabled)
            return '#c4cdd2'; // Base-30
        if (this.isError)
            return '#ef3e42'; // RHB-Red-100
        if (this.isSuccess)
            return '#6cc24a'; // Green-100
        return '#0067b1'; // RHB-Blue-100 (Helper)
    }
    get isError() {
        return !!(this.control?.touched && this.control?.errors);
    }
    get isSuccess() {
        return !!(this.control?.valid && this.control?.touched);
    }
    get isHelper() {
        return true;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.19", ngImport: i0, type: FNFieldMessage, deps: [{ token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "19.2.19", type: FNFieldMessage, isStandalone: true, selector: "fn-field-message", inputs: { field: "field", control: "control", hasFocus: "hasFocus" }, usesOnChanges: true, ngImport: i0, template: `
    <fn-label variant="p4" [color]="statusColor" [label]="getFieldMessage()"></fn-label>
  `, isInline: true, dependencies: [{ kind: "ngmodule", type: CommonModule }, { kind: "ngmodule", type: TranslateModule }, { kind: "component", type: FNLabel, selector: "fn-label", inputs: ["field", "label", "required", "className", "class", "for", "color", "variant", "statusLabel", "ngClass", "hideOptional"] }] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.19", ngImport: i0, type: FNFieldMessage, decorators: [{
            type: Component,
            args: [{
                    selector: 'fn-field-message',
                    standalone: true,
                    imports: [CommonModule, TranslateModule, FNLabel],
                    template: `
    <fn-label variant="p4" [color]="statusColor" [label]="getFieldMessage()"></fn-label>
  `,
                }]
        }], ctorParameters: () => [{ type: i0.ChangeDetectorRef }], propDecorators: { field: [{
                type: Input
            }], control: [{
                type: Input
            }], hasFocus: [{
                type: Input
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { FNFieldMessage };
