// fn-input/fesm2022/fn-input.mjs
import * as i0 from '@angular/core';
import { InjectionToken, inject, EventEmitter, signal, ViewChild, Output, Input, ChangeDetectionStrategy, Component } from '@angular/core';
import * as i2 from '@angular/common';
import { CommonModule } from '@angular/common';
import * as i1 from '@angular/forms';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import * as i3 from '@ngx-translate/core';
import { TranslateService, TranslateModule } from '@ngx-translate/core';
import { Subscription, merge } from 'rxjs';
import { startWith } from 'rxjs/operators';
import { FNLabel } from 'fn-label';
import { FNFieldMessage } from 'fn-field-message';
import { FNIconComponent } from '@arpudhabotupload/fn-icon-angular';
import { ToastService, FNToast } from 'fn-toast';

const FN_TOAST_SERVICE = new InjectionToken('FN_TOAST_SERVICE');
const DEFAULT_CURRENCY_META = [
    { code: 'USD', locale: 'en-US', symbol: '$', digit: '1.2-2' },
    { code: 'MYR', locale: 'en-MY', symbol: 'RM', digit: '1.2-2' },
    { code: 'EUR', locale: 'en-US', symbol: '€', digit: '1.2-2' },
    { code: 'GBP', locale: 'en-US', symbol: '£', digit: '1.2-2' },
    { code: 'JPY', locale: 'ja-JP', symbol: '¥', digit: '1.0-0' },
    { code: 'CNY', locale: 'zh-CN', symbol: '¥', digit: '1.2-2' },
    { code: 'INR', locale: 'en-IN', symbol: '₹', digit: '1.2-2' },
];

class FNInput {
    cdr;
    field = {};
    helperHandle;
    toastService;
    fallbackToastService = inject(ToastService, { optional: true });
    injectedToastService = inject(FN_TOAST_SERVICE, { optional: true });
    get effectiveToastService() {
        return this.toastService || this.injectedToastService || this.fallbackToastService || undefined;
    }
    currencyMeta = DEFAULT_CURRENCY_META;
    defaultLocale = 'en-US';
    form;
    valueChange = new EventEmitter();
    fieldBlur = new EventEmitter();
    translateService = inject(TranslateService);
    hasFocus = signal(false);
    textareaElement;
    // Removed local Toast State
    // Removed HttpClient injection
    subs = new Subscription();
    sizeMap = {
        extrasmall: 10,
        small: 16,
        medium: 20,
        large: 24,
        'x-large': 32,
        xxlarge: 48,
    };
    getIconSizeName(size) {
        switch (size) {
            case 16:
                return 'small';
            case 20:
                return 'medium';
            case 24:
                return 'large';
            case 32:
                return 'x-large';
            default:
                return 'medium';
        }
    }
    get isAlphanumeric() {
        return this.field.isAlphanumeric !== false;
    }
    get isEmailField() {
        return this.field.type === 'email' || this.field.name?.toLowerCase().includes('email');
    }
    get alphanumericPattern() {
        if (this.field.isAddressLine) {
            return /[^A-Za-z0-9 \-_&(),/]/g;
        }
        return /[^A-Za-z0-9 \-_&()]/g;
    }
    isDisabled = false;
    control;
    isVisible = signal(true); // Track visibility state
    helperText = '';
    // Use a counter-based approach for unique IDs (safer than Math.random())
    static idCounter = 0;
    uniqueId = `fn-input-${++FNInput.idCounter}`;
    constructor(cdr) {
        this.cdr = cdr;
    }
    ngOnInit() {
        if (this.field.hidden)
            return;
        this.helperText = this.field.helperText ?? '';
        this.initFormControl();
        if (this.field.value)
            this.control.setValue(this.field.value);
        this.setupVisibilityCondition();
        this.setupFieldMessageListener();
    }
    ngOnDestroy() {
        this.subs.unsubscribe();
    }
    setupFieldMessageListener() {
        if (!this.control)
            return;
        const val$ = this.control.valueChanges.pipe(startWith(this.control.value));
        const status$ = this.control.statusChanges.pipe(startWith(this.control.status));
        this.subs.add(merge(val$, status$).subscribe(() => {
            try {
                this.cdr.detectChanges();
            }
            catch (e) {
                console.warn('CDR detectChanges failed during field message update:', e);
                this.cdr.markForCheck();
            }
        }));
    }
    ngAfterViewInit() {
        // Access the textarea element by ViewChild or native element reference
        this.autoResizeInitial();
        if (this.field.type === 'textarea' && this.control?.value) {
            setTimeout(() => this.autoResizeInitial(), 30);
        }
        // Apply initial formatting after view is initialized
        if (this.field.type === 'number' && this.field.isCurrency && this.control?.value) {
            this.formatInitialCurrencyValue();
        }
    }
    enforceLowercase(event) {
        const input = event.target;
        const start = input.selectionStart;
        const end = input.selectionEnd;
        input.value = input.value.toLowerCase();
        if (this.control) {
            this.control.setValue(input.value, { emitEvent: false });
        }
        this.valueChange.emit({ name: this.field.name, value: input.value });
        // restore cursor position so typing is smooth
        input.setSelectionRange(start, end);
    }
    /**
     * Handle email input - allows only valid email characters and enforces lowercase.
     * Allowed: a-z, 0-9, . _ + - @
     * Rules:
     * - Only one @ allowed
     * - No consecutive dots (..)
     * - Automatic lowercase
     */
    handleEmailInput(event) {
        const input = event.target;
        const cursorPosition = input.selectionStart || 0;
        const originalValue = input.value;
        // 1. Convert to lowercase
        const lowerValue = originalValue.toLowerCase();
        // 2. Split at the first @ to handle local and domain parts separately
        const parts = lowerValue.split('@');
        let localPart = parts[0];
        let domainPart = parts.length > 1 ? parts.slice(1).join('') : null;
        // 3. Filter local part: a-z, 0-9, ., _, +, -
        // Rule: Cannot start with a dot (.)
        localPart = localPart.replaceAll(/[^a-z0-9._+-]/g, '');
        while (localPart.startsWith('.')) {
            localPart = localPart.substring(1);
        }
        while (localPart.includes('..')) {
            localPart = localPart.replaceAll('..', '.');
        }
        // 4. Filter domain part: a-z, 0-9, ., -
        // Rule: No underscores (_), no plus signs (+), cannot start with a hyphen (-)
        let filteredValue = localPart;
        if (domainPart !== null) {
            domainPart = domainPart.replaceAll(/[^a-z0-9.-]/g, '');
            while (domainPart.startsWith('-')) {
                domainPart = domainPart.substring(1);
            }
            while (domainPart.includes('..')) {
                domainPart = domainPart.replaceAll('..', '.');
            }
            filteredValue += '@' + domainPart;
        }
        if (originalValue !== filteredValue) {
            const charsRemovedBeforeCursor = this.countRemovedCharsBeforeCursor(originalValue.toLowerCase(), filteredValue, cursorPosition);
            input.value = filteredValue;
            this.control.setValue(filteredValue, { emitEvent: false });
            const newCursorPos = Math.max(0, cursorPosition - charsRemovedBeforeCursor);
            input.setSelectionRange(newCursorPos, newCursorPos);
        }
        else if (originalValue !== lowerValue) {
            // If only case changed (e.g., typed Uppercase)
            input.value = filteredValue;
            this.control.setValue(filteredValue, { emitEvent: false });
            input.setSelectionRange(cursorPosition, cursorPosition);
        }
        if (this.field.type === 'password' && this.field.feedback) {
            this.checkPasswordStrength(filteredValue);
        }
        this.valueChange.emit({ name: this.field.name, value: filteredValue });
        this.control.markAsTouched();
    }
    onFocus() {
        this.hasFocus.set(true);
    }
    formatInitialCurrencyValue() {
        const numericValue = Number(this.control.value);
        if (!Number.isNaN(numericValue)) {
            setTimeout(() => {
                const inputElement = document.getElementById(this.uniqueId);
                if (inputElement) {
                    inputElement.value = this.formatCurrencyWithCommas(numericValue);
                }
            }, 50);
        }
    }
    initFormControl() {
        const controller = this.form.controls?.[this.field.name];
        this.control = controller;
        if (this.field.value)
            this.control.setValue(this.field.value);
        if (this.control) {
            // Handle disabled state
            if (this.field.disabled ?? false) {
                this.control?.disable();
            }
            else {
                this.control?.enable();
            }
        }
    }
    autoResizeInitial() {
        const textarea = this.textareaElement?.nativeElement;
        if (textarea) {
            // Get the number of rows (default to 2 if not specified)
            const rows = this.field.rows || 1;
            // Calculate line height from computed styles
            const computedStyle = globalThis.getComputedStyle(textarea);
            const lineHeight = Number.parseFloat(computedStyle.lineHeight) || 32; // Default to 24px if not set
            // Calculate minimum height based on rows
            const minHeight = rows * lineHeight;
            // Reset height to auto to get accurate scrollHeight
            textarea.style.height = 'auto';
            // Use the larger of scrollHeight or minHeight
            const newHeight = Math.max(textarea.scrollHeight, minHeight);
            textarea.style.height = newHeight + 'px';
            textarea.style.overflow = 'hidden';
        }
    }
    handleTextArea(event) {
        const element = event.target;
        const cursorPosition = element.selectionStart;
        const originalValue = element.value;
        // Apply alphanumeric filtering if enabled
        if (this.field.type === 'textarea' && this.isAlphanumeric) {
            // Only allow characters based on the selected pattern
            const alphanumericPattern = this.alphanumericPattern;
            let filteredValue = originalValue.replaceAll(alphanumericPattern, '');
            // Prevent leading spaces
            if (filteredValue.startsWith(' ')) {
                filteredValue = filteredValue.trimStart();
            }
            // Replace multiple consecutive spaces with single space
            filteredValue = filteredValue.replaceAll(/\s{2,}/g, ' ');
            if (originalValue !== filteredValue) {
                // Calculate the number of characters removed before the cursor position
                const charsRemovedBeforeCursor = originalValue.length - filteredValue.length;
                element.value = filteredValue;
                this.control.setValue(filteredValue, { emitEvent: false });
                // Restore cursor position - account for removed characters
                const newCursorPos = Math.max(0, (cursorPosition || 0) - charsRemovedBeforeCursor);
                element.setSelectionRange(newCursorPos, newCursorPos);
            }
            // increase height
            this.autoResizeInitial();
            this.valueChange.emit({ name: this.field.name, value: filteredValue });
        }
        else {
            let filteredValue = originalValue;
            // Strict filtering if !isAlphanumeric
            if (!this.isAlphanumeric) {
                console.log('isAlphanumeric', this.isAlphanumeric);
                const allowedPattern = /[^a-zA-Z0-9\s(){}[\];,."='+\-*/<>!&|%_@$?:]/g;
                filteredValue = filteredValue.replaceAll(allowedPattern, '');
            }
            // Prevent leading spaces
            if (filteredValue.startsWith(' ')) {
                filteredValue = filteredValue.trimStart();
            }
            // Replace multiple consecutive spaces with single space
            filteredValue = filteredValue.replaceAll(/\s{2,}/g, ' ');
            if (originalValue !== filteredValue) {
                // Calculate the number of characters removed before the cursor position
                const charsRemovedBeforeCursor = originalValue.length - filteredValue.length;
                element.value = filteredValue;
                this.control.setValue(filteredValue, { emitEvent: false });
                // Restore cursor position - account for removed characters
                const newCursorPos = Math.max(0, (cursorPosition || 0) - charsRemovedBeforeCursor);
                element.setSelectionRange(newCursorPos, newCursorPos);
            }
            // increase height
            this.autoResizeInitial();
            this.valueChange.emit({ name: this.field.name, value: filteredValue });
        }
        this.control.markAsTouched();
    }
    togglePasswordVisibility() {
        this.showPassword = !this.showPassword;
    }
    handleInput(event) {
        const target = event.target;
        const cursorPosition = target.selectionStart;
        const originalValue = target.value;
        let filteredValue = originalValue;
        // IF NOT alphanumeric, user wants strict filtering:
        // "only accept tihs.) {} [] ; , . " ' = + - * / < > ! & | % &_( )-/.;@!, $% : dont allow emoji"
        // Interpretation: Allow alphanumeric + specified special chars. Ban emojis.
        // Actually, the user said: "if isAlphanumeric false only accept..."
        // So if isAlphanumeric is TRUE, we use the existing handleAlphanumericInput logic (or similar).
        // The current logic in handleInput is generic.
        // The existing code has a separate `handleAlphanumericInput` method but it's not called here?
        // Wait, the template likely calls `handleAlphanumericInput` if `isAlphanumeric` is true, or `handleInput` otherwise?
        // Let's check the template.
        // BUT assuming `handleInput` is the generic handler:
        if (!this.isAlphanumeric) {
            // Regex to allow: a-z, A-Z, 0-9, and: ) {} [] ; , . " ' = + - * / < > ! & | % _ ( - @ $ ? :
            // Note: User list: `.` `"` `'` `?` (wait, ? not in list?)
            // User list: `) {} [] ; , . " ' = + - * / < > ! & | % &_( )-/.;@!, $%`
            // Uniques: ) { } [ ] ; , . " ' = + - * / < > ! & | % _ ( @ $
            // (The `&` is repeated, `.` repeated, `,` repeated etc.)
            // I will create a regex that matches anything NOT in this list and replace with empty string.
            // [^a-zA-Z0-9\s(){}[\];,."='+\-*/<>!&|%_@$]
            // Note: `*`, `+`, `?` etc need escaping in regex if outside [], but inside [] they are literals mostly.
            // `-` needs to be last or escaped. `]` needs escape. `^` needs escape if first? No, invalid here.
            // Let's verify valid chars:
            // a-z A-Z 0-9
            // space
            // ) { } [ ] ; , . " ' = + - * / < > ! & | % _ ( @ $ :
            const allowedPattern = /[^a-zA-Z0-9\s(){}[\];,."='+\-*/<>!&|%_@$?:]/g;
            filteredValue = filteredValue.replaceAll(allowedPattern, '');
        }
        // Prevent leading spaces
        if (filteredValue.startsWith(' ')) {
            filteredValue = filteredValue.trimStart();
        }
        // replaceAll multiple consecutive spaces with single space
        filteredValue = filteredValue.replaceAll(/\s{2,}/g, ' ');
        if (originalValue !== filteredValue) {
            // Calculate the number of characters removed before the cursor position
            const charsRemovedBeforeCursor = originalValue.length - filteredValue.length;
            target.value = filteredValue;
            this.control.setValue(filteredValue, { emitEvent: false });
            // Restore cursor position - account for removed characters
            const newCursorPos = Math.max(0, (cursorPosition || 0) - charsRemovedBeforeCursor);
            target.setSelectionRange(newCursorPos, newCursorPos);
        }
        // Trailing spaces will be trimmed on blur, not during input
        if (this.field.type === 'password' && this.field.feedback) {
            this.checkPasswordStrength(filteredValue);
        }
        this.valueChange.emit({ name: this.field.name, value: filteredValue });
        this.control.markAsTouched();
    }
    /**
     * Handle alphanumeric input - allows only letters, numbers, and specific special characters
     * Allowed: A-Z, a-z, 0-9, space, hyphen, underscore, ampersand, and parentheses
     * Also prevents leading spaces and multiple consecutive spaces
     */
    handleAlphanumericInput(event) {
        const target = event.target;
        const cursorPosition = target.selectionStart;
        const originalValue = target.value;
        // Only allow characters based on the selected pattern
        const alphanumericPattern = this.alphanumericPattern;
        let filteredValue = originalValue.replaceAll(alphanumericPattern, '');
        // Prevent leading spaces
        if (filteredValue.startsWith(' ')) {
            filteredValue = filteredValue.trimStart();
        }
        // Replace multiple consecutive spaces with single space
        filteredValue = filteredValue.replaceAll(/\s{2,}/g, ' ');
        if (originalValue !== filteredValue) {
            // Calculate the number of characters removed before the cursor position
            const charsRemovedBeforeCursor = originalValue.length - filteredValue.length;
            target.value = filteredValue;
            this.control.setValue(filteredValue, { emitEvent: false });
            // Restore cursor position - account for removed characters
            const newCursorPos = Math.max(0, (cursorPosition || 0) - charsRemovedBeforeCursor);
            target.setSelectionRange(newCursorPos, newCursorPos);
        }
        if (this.field.type === 'password' && this.field.feedback) {
            this.checkPasswordStrength(filteredValue);
        }
        this.valueChange.emit({ name: this.field.name, value: filteredValue });
        this.control.markAsTouched();
    }
    handleBlur(e, maxDecimals = 2) {
        this.hasFocus.set(false);
        const target = e.target;
        const cleanValueString = target.value;
        if (!cleanValueString) {
            this.fieldBlur.emit({ name: this.field.name, value: '' });
            this.cdr.detectChanges();
            return;
        }
        // Delegate to type-specific handlers
        if (this.field.type === 'text' || this.field.type === 'textarea') {
            this.handleTextFieldBlur(target, cleanValueString);
        }
        else if (this.field.type === 'number') {
            this.handleNumberFieldBlur(target, cleanValueString, maxDecimals);
        }
        // Emit final value
        this.fieldBlur.emit({
            name: this.field.name,
            value: target.value,
        });
        this.cdr.detectChanges();
    }
    /**
     * Handle blur for text and textarea fields - trim trailing spaces
     */
    handleTextFieldBlur(target, cleanValueString) {
        const trimmed = cleanValueString.trim();
        if (cleanValueString !== trimmed) {
            target.value = trimmed;
            this.control.setValue(trimmed, { emitEvent: false });
        }
    }
    /**
     * Handle blur for number fields - validate and format
     */
    handleNumberFieldBlur(target, cleanValueString, maxDecimals) {
        // Remove commas for parsing/validation
        const numString = cleanValueString.replaceAll(',', '');
        // Check if the string is a valid number format
        const isValidNumberFormat = /^-?\d+(\.\d+)?$/.test(numString);
        // If not a valid number format, clear the field
        if (!isValidNumberFormat) {
            target.value = '';
            this.control.setValue('');
            return;
        }
        // Format currency fields or set raw value for non-currency
        if (this.field.type === 'number' && this.field.isCurrency) {
            this.processCurrencyValue(target, numString, maxDecimals);
        }
        else {
            // Non-currency: preserve the raw string value
            target.value = numString;
        }
    }
    /**
     * Process currency value - handle formatting and decimal places
     */
    processCurrencyValue(target, numString, maxDecimals) {
        // Handle negative values - convert to positive
        let processedValue = numString.startsWith('-') ? numString.substring(1) : numString;
        // Handle edge case of "-0" or just "-"
        if (processedValue === '' || processedValue === '0') {
            processedValue = '0';
        }
        // Format the value with proper decimal places
        const cleanValueString = this.formatDecimalPlaces(processedValue, maxDecimals);
        const formattedValue = this.formatCurrencyWithCommas(cleanValueString, maxDecimals);
        // Update the control value with the numeric string (not formatted)
        this.control.setValue(cleanValueString, { emitEvent: false });
        // Trigger validators and mark as touched
        this.control.updateValueAndValidity();
        this.control.markAsTouched();
        // Set display value with commas after Angular's form control update
        setTimeout(() => {
            target.value = formattedValue;
        }, 0);
    }
    /**
     * Format decimal places - pad or truncate to maxDecimals
     */
    formatDecimalPlaces(value, maxDecimals) {
        let [intPart, decPart] = value.includes('.') ? value.split('.') : [value, ''];
        // Pad or truncate decimal part to maxDecimals
        if (decPart.length < maxDecimals) {
            decPart = decPart.padEnd(maxDecimals, '0');
        }
        else if (decPart.length > maxDecimals) {
            decPart = decPart.slice(0, maxDecimals);
        }
        // Build the clean value string (without commas)
        return maxDecimals > 0 ? `${intPart}.${decPart}` : intPart;
    }
    handleNumberInput(event, maxDecimals = 2) {
        const target = event.target;
        const originalValue = target.value;
        const cursorPosition = target.selectionStart ?? 0;
        // Step 1: Filter and format the input value
        let value = this.filterNumericInput(target.value, maxDecimals);
        // Step 2: Update cursor position if value changed
        this.updateCursorPositionAfterFilter(target, originalValue, value, cursorPosition);
        // Step 3: Update form control and validate
        this.updateFormControlValue(value, maxDecimals);
        // Step 4: Emit changes
        this.valueChange.emit({ name: this.field.name, value: target.value });
        this.control.markAsTouched();
    }
    /**
     * Filter numeric input to only allow digits and decimal point
     * Ensures only one decimal point and enforces max decimal places
     */
    filterNumericInput(value, maxDecimals) {
        // Remove ALL non-numeric characters except decimal point
        let filtered = value.replaceAll(/[^0-9.]/g, '');
        // Ensure only one decimal point
        filtered = this.filterDecimalPoints(filtered);
        // Restrict to maxDecimals decimal places if decimal exists
        filtered = this.enforceDecimalLimit(filtered, maxDecimals);
        return filtered;
    }
    /**
     * Ensure only one decimal point exists in the value
     */
    filterDecimalPoints(value) {
        const parts = value.split('.');
        if (parts.length > 2) {
            return parts[0] + '.' + parts.slice(1).join('');
        }
        return value;
    }
    /**
     * Enforce maximum decimal places limit
     */
    enforceDecimalLimit(value, maxDecimals) {
        if (!value.includes('.')) {
            return value;
        }
        const [integerPart, decimalPart] = value.split('.');
        if (decimalPart && decimalPart.length > maxDecimals) {
            return integerPart + '.' + decimalPart.slice(0, maxDecimals);
        }
        return value;
    }
    /**
     * Update cursor position after filtering input
     */
    updateCursorPositionAfterFilter(target, originalValue, newValue, cursorPosition) {
        if (target.value === newValue) {
            return;
        }
        const charsRemovedBeforeCursor = this.countRemovedCharsBeforeCursor(originalValue, newValue, cursorPosition);
        target.value = newValue;
        const newPosition = Math.max(0, cursorPosition - charsRemovedBeforeCursor);
        target.setSelectionRange(newPosition, newPosition);
    }
    /**
     * Update form control value and perform validation
     */
    updateFormControlValue(value, maxDecimals) {
        this.control.setValue(value, { emitEvent: false });
        this.validateIntegerDigits(value, maxDecimals);
        this.clearNumericError();
    }
    /**
     * Count how many characters were removed before the cursor position
     * This helps maintain correct cursor position after filtering input
     */
    countRemovedCharsBeforeCursor(original, filtered, cursorPos) {
        let removedCount = 0;
        let filteredIndex = 0;
        for (let i = 0; i < cursorPos && i < original.length; i++) {
            if (filteredIndex < filtered.length && original[i] === filtered[filteredIndex]) {
                filteredIndex++;
            }
            else {
                removedCount++;
            }
        }
        return removedCount;
    }
    /**
     * Handle paste event for number inputs
     * Prevents pasting of non-numeric characters and enforces digit limits
     */
    handleNumberPaste(event, maxDecimals = 2) {
        event.preventDefault();
        const target = event.target;
        const pastedText = event.clipboardData?.getData('text') || '';
        const isCurrency = this.field.type === 'number' && !!this.field.isCurrency;
        const filteredText = isCurrency
            ? this.getFilteredCurrencyPastedText(pastedText)
            : this.getFilteredIntegerPastedText(pastedText);
        const finalValue = this.buildFinalPasteValue(target, filteredText, isCurrency, maxDecimals);
        // Update the input
        target.value = finalValue;
        this.control.setValue(finalValue, { emitEvent: false });
        this.validateIntegerDigits(finalValue, maxDecimals);
        this.clearNumericError();
        this.valueChange.emit({ name: this.field.name, value: finalValue });
        this.control.markAsTouched();
        this.setCursorAfterPaste(target, filteredText, finalValue);
    }
    /**
     * Filter pasted text for currency input (allows digits and decimal point)
     */
    getFilteredCurrencyPastedText(pastedText) {
        return pastedText.replaceAll(/[^0-9.]/g, '');
    }
    /**
     * Filter pasted text for integer-only input (allows digits only)
     */
    getFilteredIntegerPastedText(pastedText) {
        return pastedText.replaceAll(/\D/g, '');
    }
    /**
     * Build final value after paste operation
     */
    buildFinalPasteValue(target, filteredText, isCurrency, maxDecimals) {
        const currentValue = target.value;
        const start = target.selectionStart;
        const end = target.selectionEnd;
        let finalValue;
        // If selection API is not supported or no selection, replace entire content
        if (start === null || end === null || (start === 0 && end === 0)) {
            finalValue = filteredText;
        }
        else {
            finalValue = currentValue.substring(0, start) + filteredText + currentValue.substring(end);
        }
        return isCurrency ? this.applyDecimalRestrictions(finalValue, maxDecimals) : finalValue;
    }
    /**
     * Apply decimal restrictions for currency fields
     */
    applyDecimalRestrictions(value, maxDecimals) {
        let result = value;
        // Ensure only one decimal point
        const parts = result.split('.');
        if (parts.length > 2) {
            result = parts[0] + '.' + parts.slice(1).join('');
        }
        // Restrict to maxDecimals decimal places
        if (result.includes('.')) {
            const [intPart, decPart] = result.split('.');
            if (decPart.length > maxDecimals) {
                result = intPart + '.' + decPart.slice(0, maxDecimals);
            }
        }
        return result;
    }
    /**
     * Validate integer digit limit and set/clear errors
     */
    validateIntegerDigits(value, maxDecimals) {
        if (this.field.type !== 'number' || !this.field.maxIntegerDigits) {
            return;
        }
        const maxIntDigits = this.field.maxIntegerDigits ?? 15;
        const [intPart] = value.includes('.') ? value.split('.') : [value];
        if (intPart.length > maxIntDigits) {
            this.setMaxIntegerDigitsError(intPart.length, maxIntDigits, maxDecimals);
        }
        else {
            this.clearMaxIntegerDigitsError();
        }
    }
    /**
     * Set max integer digits error
     */
    setMaxIntegerDigitsError(actual, max, maxDecimals) {
        const currentErrors = this.control.errors || {};
        this.control.setErrors({
            ...currentErrors,
            maxIntegerDigits: {
                actual,
                max,
                maxDecimals,
                message: `Please shorten your input to ${max} whole digits and ${maxDecimals} decimal or fewer.`,
            },
        });
    }
    /**
     * Clear max integer digits error if present
     */
    clearMaxIntegerDigitsError() {
        if (!this.control.hasError('maxIntegerDigits')) {
            return;
        }
        const errors = { ...this.control.errors };
        delete errors['maxIntegerDigits'];
        this.control.setErrors(Object.keys(errors).length ? errors : null);
    }
    /**
     * Clear numeric error if present
     */
    clearNumericError() {
        if (!this.control.hasError('numeric')) {
            return;
        }
        const errors = { ...this.control.errors };
        delete errors['numeric'];
        this.control.setErrors(Object.keys(errors).length ? errors : null);
    }
    /**
     * Set cursor position after paste operation
     */
    setCursorAfterPaste(target, filteredText, finalValue) {
        // Input type="number" does not support selection APIs
        if (target.type === 'number') {
            return;
        }
        const start = target.selectionStart || 0;
        const newCursorPos = start + filteredText.length;
        const actualPos = Math.min(newCursorPos, finalValue.length);
        target.setSelectionRange(actualPos, actualPos);
    }
    // Format number with comma separators
    // Uses string-based formatting for large numbers to preserve precision
    formatCurrencyWithCommas(value, maxDecimals = 2) {
        if (this.field.type !== 'number' || !this.field.isCurrency)
            return value.toString();
        const strValue = value.toString();
        // Return original value for empty or invalid inputs
        if (!strValue || strValue === 'NaN')
            return '';
        // Use string-based formatting to preserve precision for large numbers
        const [integerPart, decimalPart] = strValue.includes('.')
            ? strValue.split('.')
            : [strValue, ''];
        // Format integer part with comma separators (every 3 digits from right)
        // Using iterative approach instead of regex to avoid ReDoS concerns
        let formattedInteger = '';
        const len = integerPart.length;
        for (let i = 0; i < len; i++) {
            if (i > 0 && (len - i) % 3 === 0) {
                formattedInteger += ',';
            }
            formattedInteger += integerPart[i];
        }
        // Format decimal part (pad or truncate to maxDecimals)
        let formattedDecimal = decimalPart || '';
        if (formattedDecimal.length < maxDecimals) {
            formattedDecimal = formattedDecimal.padEnd(maxDecimals, '0');
        }
        else if (formattedDecimal.length > maxDecimals) {
            formattedDecimal = formattedDecimal.slice(0, maxDecimals);
        }
        return maxDecimals > 0 ? `${formattedInteger}.${formattedDecimal}` : formattedInteger;
    }
    handleNumberKeydown(event, isCurrency) {
        if (isCurrency)
            return;
        // Allow: Backspace, Delete, Tab, Arrow keys
        if (['Backspace', 'Delete', 'Tab', 'ArrowLeft', 'ArrowRight'].includes(event.key)) {
            return;
        }
        // Allow clipboard operations (Ctrl+V, Ctrl+C, Ctrl+X, Cmd+V, Cmd+C, Cmd+X)
        // and selection (Ctrl+A, Cmd+A)
        if (event.ctrlKey || event.metaKey) {
            return;
        }
        // Block everything that is not 0-9
        if (!/^\d$/.test(event.key)) {
            event.preventDefault();
        }
    }
    // Get locale from currency code using currencyMeta
    getLocaleFromCurrency(currencyCode) {
        const currencyObj = this.currencyMeta.find((c) => c.code === currencyCode);
        return currencyObj?.locale || this.defaultLocale;
    }
    showPassword = false;
    passwordStrengthLabel = '';
    strengthClass = '';
    passwordStrengthPercent = 0;
    isPasswordFocused = false;
    onPasswordFocus() {
        this.hasFocus.set(true);
        this.isPasswordFocused = true;
        this.cdr.detectChanges();
    }
    onPasswordBlur() {
        this.isPasswordFocused = false;
        this.cdr.detectChanges();
    }
    handlePasswordBlur(e) {
        this.handleBlur(e); // Original blur functionality (sets hasFocus to false)
        this.onPasswordBlur(); // Hide password feedback
    }
    checkPasswordStrength(value) {
        if (!value || this.field.type !== 'password') {
            this.passwordStrengthLabel = '';
            this.strengthClass = '';
            this.passwordStrengthPercent = 0;
            return;
        }
        let score = 0;
        if (value.length >= 8)
            score++;
        if (/[A-Z]/.test(value))
            score++;
        if (/\d/.test(value))
            score++;
        if (/[\W_]/.test(value))
            score++;
        if (score <= 1) {
            this.passwordStrengthLabel = this.field.weakLabel || 'Weak';
            this.strengthClass = 'text-[var(--RHB-Red-100)]';
            this.passwordStrengthPercent = 25;
        }
        else if (score === 2 || score === 3) {
            this.passwordStrengthLabel = this.field.mediumLabel || 'Medium';
            this.strengthClass = 'text-[var(--Orange-100)]';
            this.passwordStrengthPercent = score === 2 ? 50 : 75;
        }
        else {
            this.passwordStrengthLabel = this.field.strongLabel || 'Strong';
            this.strengthClass = 'text-[var(--Green-100)]';
            this.passwordStrengthPercent = 100;
        }
    }
    increment() {
        if (this.field.type !== 'number')
            return;
        const step = this.field.step || 1;
        const current = Number(this.control.value) || 0;
        let nextValue = current + step;
        if (typeof this.field.max === 'number') {
            nextValue = Math.min(nextValue, this.field.max);
        }
        this.control.setValue(nextValue);
        this.control.markAsTouched();
        this.valueChange.emit({ name: this.field.name, value: nextValue });
        // Update display value for currency inputs
        if (this.field.isCurrency) {
            setTimeout(() => {
                const inputElement = document.getElementById(this.uniqueId);
                if (inputElement) {
                    inputElement.value = this.formatCurrencyWithCommas(nextValue);
                }
            });
        }
    }
    decrement() {
        if (this.field.type !== 'number')
            return;
        const step = this.field.step || 1;
        const current = Number(this.control.value) || 0;
        let nextValue = current - step;
        // Ensure minimum value is 0 (no negative values)
        const minValue = Math.max(this.field.min || 0, 0);
        nextValue = Math.max(nextValue, minValue);
        this.control.setValue(nextValue);
        this.control.markAsTouched();
        this.valueChange.emit({ name: this.field.name, value: nextValue });
        // Update display value for currency inputs
        if (this.field.isCurrency) {
            setTimeout(() => {
                const inputElement = document.getElementById(this.uniqueId);
                if (inputElement) {
                    inputElement.value = this.formatCurrencyWithCommas(nextValue);
                }
            });
        }
    }
    copyText(e) {
        e.stopPropagation();
        if (!this.control.value)
            return;
        navigator.clipboard.writeText(this.control.value).then(() => {
            const service = this.effectiveToastService;
            if (service) {
                // Position set to top as requested
                service.success('Copied to clipboard', 'Success', 2000);
            }
        });
    }
    prefixClick(e) {
        if (this.field.prefix?.onClick) {
            this.field.prefix.onClick(e);
        }
    }
    suffixClick(e) {
        if (this.field.suffix?.onClick) {
            this.field.suffix.onClick(e);
        }
    }
    /**
     * Set up visibility condition listener
     * Watches the dependent field and shows/hides this field based on conditions
     */
    setupVisibilityCondition() {
        if (!this.field.visibilityCondition || !this.form)
            return;
        const { dependsOn } = this.field.visibilityCondition;
        const dependentControl = this.form.get(dependsOn);
        if (!dependentControl) {
            return;
        }
        // Initial visibility check
        this.updateVisibility(dependentControl.value);
        // Subscribe to value changes
        dependentControl.valueChanges.subscribe((value) => {
            this.updateVisibility(value);
        });
    }
    /**
     * Update field visibility based on dependent field value
     */
    updateVisibility(dependentValue) {
        if (!this.field.visibilityCondition)
            return;
        const { showWhen, hideWhen } = this.field.visibilityCondition;
        let shouldBeVisible = false;
        // Check showWhen condition
        if (showWhen !== undefined) {
            if (Array.isArray(showWhen)) {
                shouldBeVisible = showWhen.includes(dependentValue);
            }
            else {
                shouldBeVisible = dependentValue === showWhen;
            }
        }
        // Check hideWhen condition (overrides showWhen)
        if (hideWhen !== undefined) {
            const shouldHide = Array.isArray(hideWhen)
                ? hideWhen.includes(dependentValue)
                : dependentValue === hideWhen;
            if (shouldHide) {
                shouldBeVisible = false;
            }
        }
        // Update visibility
        this.isVisible.set(shouldBeVisible);
        // Only clear value when hiding if there's no initial value
        // This preserves data when editing existing records
        if (!shouldBeVisible && this.control && !this.control.value) {
            this.control.setValue(null);
            this.control.markAsUntouched();
            this.control.updateValueAndValidity();
        }
        this.cdr.detectChanges();
    }
    // Removed local toast helpers
    showFormFieldMessage(control, helperText) {
        if (this.helperHandle) {
            return this.helperHandle.showFormFieldMessage(control, helperText);
        }
        return true;
    }
    getTranslatedLabel(label) {
        if (!label)
            return '';
        try {
            const labelStr = label.toString();
            return (labelStr
                .split('::')
                .map((part) => part.trim())
                .filter((part) => !!part)
                .map((part) => this.translateService.instant(part))
                .join(' ') + this.getEndSymbol(labelStr));
        }
        catch (e) {
            console.warn('Translation failed in getTranslatedLabel:', label, e);
            return label;
        }
    }
    getEndSymbol(label) {
        return label.toString().includes('required') && label.toString().includes('FNFieldMessage')
            ? '.'
            : '';
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.19", ngImport: i0, type: FNInput, deps: [{ token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "19.2.19", type: FNInput, isStandalone: true, selector: "fn-input", inputs: { field: "field", helperHandle: "helperHandle", toastService: "toastService", currencyMeta: "currencyMeta", defaultLocale: "defaultLocale", form: "form" }, outputs: { valueChange: "valueChange", fieldBlur: "fieldBlur" }, viewQueries: [{ propertyName: "textareaElement", first: true, predicate: ["fnTextarea"], descendants: true }], ngImport: i0, template: "@if (field.name && isVisible() && !field.hidden) {\r\n  <div class=\"fn-input-container\">\r\n    @if (field.type! !== 'hidden') {\r\n      <fn-label\r\n        [for]=\"uniqueId\"\r\n        [label]=\"field.label\"\r\n        [required]=\"field.required || false\"\r\n        [hideOptional]=\"field.hideOptional || false\"\r\n        [color]=\"field.color || '#03182b'\"\r\n        [variant]=\"field.labelVariant || 'p1'\"\r\n        [statusLabel]=\"field.statusLabel\"\r\n      ></fn-label>\r\n    }\r\n\r\n    <ng-container>\r\n      @switch (field.type) {\r\n        <!-- Textarea Field -->\r\n        @case ('textarea') {\r\n          <textarea\r\n            #fnTextarea\r\n            style=\"resize: none\"\r\n            [name]=\"field.name\"\r\n            [id]=\"uniqueId\"\r\n            [required]=\"field.required || false\"\r\n            [placeholder]=\"field.placeholder || '' | translate\"\r\n            [disabled]=\"field.disabled || false\"\r\n            [rows]=\"field.rows || 1\"\r\n            [readonly]=\"field.readOnly || false\"\r\n            [formControl]=\"control\"\r\n            (input)=\"handleTextArea($event)\"\r\n            (blur)=\"handleBlur($event)\"\r\n            class=\"fn-input-field\"\r\n            [ngClass]=\"{\r\n              disabled: control.disabled || field.readOnly,\r\n              error: control.touched && control.errors,\r\n              success: field['hasSuccessBorder'] && control.valid && !control.errors,\r\n            }\"\r\n            [ngStyle]=\"{\r\n              color: field.valueColor || 'inherit',\r\n              'font-size': field.valueSize || '',\r\n            }\"\r\n            (focus)=\"onFocus()\"\r\n          ></textarea>\r\n        }\r\n\r\n        <!-- Password Field -->\r\n        @case ('password') {\r\n          <div class=\"relative w-full\">\r\n            <div\r\n              class=\"absolute left-0 top-1/2 -translate-y-1/2 pl-3 flex items-center gap-2\"\r\n              style=\"z-index: 1\"\r\n            >\r\n              @if (field.prefix?.icon) {\r\n                <fn-icon\r\n                  [name]=\"field.prefix?.icon || ''\"\r\n                  [variant]=\"'Line'\"\r\n                  [size]=\"'medium'\"\r\n                  [color]=\"control.disabled ? '#c4cdd2' : '#03182b'\"\r\n                ></fn-icon>\r\n              }\r\n              @if (field.prefix?.text) {\r\n                <span\r\n                  class=\"text-sm font-medium whitespace-nowrap\"\r\n                  [ngStyle]=\"{ color: control.disabled ? '#c4cdd2' : '#03182b' }\"\r\n                  (click)=\"prefixClick($event)\"\r\n                  (keydown.enter)=\"prefixClick($event)\"\r\n                  [attr.tabindex]=\"field.prefix?.onClick ? 0 : null\"\r\n                  [class.cursor-pointer]=\"!!field.prefix?.onClick\"\r\n                  >{{ field.prefix!.text }}</span\r\n                >\r\n              }\r\n            </div>\r\n            <input\r\n              [type]=\"showPassword ? 'text' : 'password'\"\r\n              [id]=\"uniqueId\"\r\n              [required]=\"field.required || false\"\r\n              [disabled]=\"field.disabled || false\"\r\n              [formControl]=\"control\"\r\n              [placeholder]=\"field.placeholder || '' | translate\"\r\n              [readonly]=\"field.readOnly || false\"\r\n              (input)=\"isAlphanumeric ? handleAlphanumericInput($event) : handleInput($event)\"\r\n              (focus)=\"onPasswordFocus()\"\r\n              (blur)=\"handlePasswordBlur($event)\"\r\n              class=\"fn-input-field overflow-hidden text-ellipsis\"\r\n              [ngClass]=\"{\r\n                disabled: control.disabled || field.readOnly,\r\n                error: control.touched && control.errors,\r\n                success: field['hasSuccessBorder'] && control.valid,\r\n                'pl-10': field.prefix?.icon && !field.prefix?.text,\r\n                'has-prefix-text': !!field.prefix?.text,\r\n                'pr-12':\r\n                  (field.toggleMask && !field.isCopyText) ||\r\n                  (!field.toggleMask && field.isCopyText),\r\n                'pr-24': field.toggleMask && field.isCopyText && !field.hasGenerateKey,\r\n                'pr-32': field.toggleMask && field.isCopyText && field.hasGenerateKey,\r\n              }\"\r\n              [ngStyle]=\"{\r\n                color: field.valueColor || 'inherit',\r\n                'font-size': field.valueSize || '',\r\n              }\"\r\n            />\r\n\r\n            <div class=\"absolute right-0 top-1/2 -translate-y-1/2 flex items-center pr-3 gap-2\">\r\n              @if (field.toggleMask && !field.isCopyText) {\r\n                <span\r\n                  class=\"cursor-pointer\"\r\n                  (click)=\"togglePasswordVisibility()\"\r\n                  (keydown)=\"togglePasswordVisibility()\"\r\n                >\r\n                  <fn-icon\r\n                    [name]=\"showPassword ? 'eye-open' : 'eye-close'\"\r\n                    [variant]=\"'Line'\"\r\n                    [size]=\"'large'\"\r\n                    [color]=\"'#03182b'\"\r\n                  ></fn-icon>\r\n                </span>\r\n              } @else if (field.isCopyText && !field.toggleMask) {\r\n                <span\r\n                  class=\"cursor-pointer\"\r\n                  (click)=\"copyText($event)\"\r\n                  (keydown)=\"copyText($event)\"\r\n                >\r\n                  <fn-icon [name]=\"'two-square'\" [color]=\"'#03182b'\"></fn-icon>\r\n                </span>\r\n                @if (field.hasGenerateKey) {\r\n                  <span\r\n                    class=\"cursor-pointer\"\r\n                    (click)=\"field.onGenerateKey && field.onGenerateKey()\"\r\n                    (keydown)=\"field.onGenerateKey && field.onGenerateKey()\"\r\n                  >\r\n                    <fn-icon\r\n                      [name]=\"'round-arrow-top-left'\"\r\n                      [variant]=\"'Line'\"\r\n                      [size]=\"'large'\"\r\n                      [color]=\"'#03182b'\"\r\n                    ></fn-icon>\r\n                  </span>\r\n                }\r\n              } @else if (field.toggleMask && field.isCopyText) {\r\n                <span\r\n                  class=\"cursor-pointer\"\r\n                  (click)=\"togglePasswordVisibility()\"\r\n                  (keydown)=\"togglePasswordVisibility()\"\r\n                >\r\n                  <fn-icon\r\n                    [name]=\"showPassword ? 'eye-open' : 'eye-close'\"\r\n                    [variant]=\"'Line'\"\r\n                    [size]=\"'large'\"\r\n                    [color]=\"'#03182b'\"\r\n                  ></fn-icon>\r\n                </span>\r\n                <span\r\n                  class=\"cursor-pointer\"\r\n                  (click)=\"copyText($event)\"\r\n                  (keydown)=\"copyText($event)\"\r\n                >\r\n                  <fn-icon [name]=\"'two-square'\" [color]=\"'#03182b'\"></fn-icon>\r\n                </span>\r\n                @if (field.hasGenerateKey) {\r\n                  <span\r\n                    class=\"cursor-pointer\"\r\n                    (click)=\"field.onGenerateKey && field.onGenerateKey()\"\r\n                    (keydown)=\"field.onGenerateKey && field.onGenerateKey()\"\r\n                    [ngClass]=\"{ 'opacity-50 cursor-not-allowed': !control.value }\"\r\n                  >\r\n                    <fn-icon [name]=\"'round-arrow-top-left'\" [color]=\"'#03182b'\"></fn-icon>\r\n                  </span>\r\n                }\r\n              }\r\n            </div>\r\n          </div>\r\n\r\n          @if (field.feedback && isPasswordFocused) {\r\n            <div class=\"mt-2 space-y-1\">\r\n              <div class=\"flex items-center justify-between text-xs\">\r\n                <span class=\"font-medium\" [ngClass]=\"strengthClass\">\r\n                  {{ passwordStrengthLabel | translate }}\r\n                </span>\r\n                <div class=\"flex items-center gap-1\">\r\n                  <fn-icon\r\n                    [name]=\"passwordStrengthPercent > 50 ? 'check-circle' : 'info-circle'\"\r\n                    [variant]=\"'Line'\"\r\n                    [size]=\"'extrasmall'\"\r\n                    [color]=\"\r\n                      passwordStrengthPercent <= 25\r\n                        ? '#ef3e42'\r\n                        : passwordStrengthPercent <= 75\r\n                          ? '#ff9f00'\r\n                          : '#6cc24a'\r\n                    \"\r\n                  ></fn-icon>\r\n                </div>\r\n              </div>\r\n              <div class=\"h-1.5 w-full bg-[#eef0f2] rounded-full overflow-hidden\">\r\n                <div\r\n                  class=\"h-full transition-all duration-300 rounded-full\"\r\n                  [ngClass]=\"{\r\n                    'bg-[#ef3e42]': passwordStrengthPercent <= 25,\r\n                    'bg-[#ff9f00]': passwordStrengthPercent > 25 && passwordStrengthPercent <= 75,\r\n                    'bg-[#6cc24a]': passwordStrengthPercent > 75,\r\n                  }\"\r\n                  [style.width.%]=\"passwordStrengthPercent\"\r\n                ></div>\r\n              </div>\r\n            </div>\r\n          }\r\n        }\r\n\r\n        <!-- Number Field -->\r\n        @case ('number') {\r\n          <div class=\"relative w-full\">\r\n            @if (field.prefix?.icon || field.prefix?.text) {\r\n              <div\r\n                class=\"absolute left-0 top-1/2 -translate-y-1/2 pl-3 flex items-center gap-2\"\r\n                style=\"z-index: 1\"\r\n              >\r\n                @if (field.prefix?.icon) {\r\n                  <fn-icon\r\n                    [name]=\"field.prefix?.icon || ''\"\r\n                    [variant]=\"'Line'\"\r\n                    [size]=\"'medium'\"\r\n                    [color]=\"control.disabled ? '#c4cdd2' : '#03182b'\"\r\n                  ></fn-icon>\r\n                }\r\n                @if (field.prefix?.text) {\r\n                  <span\r\n                    class=\"text-sm font-medium whitespace-nowrap\"\r\n                    [ngStyle]=\"{ color: control.disabled ? '#c4cdd2' : '#03182b' }\"\r\n                    (click)=\"prefixClick($event)\"\r\n                    (keydown.enter)=\"prefixClick($event)\"\r\n                    [attr.tabindex]=\"field.prefix?.onClick ? 0 : null\"\r\n                    [class.cursor-pointer]=\"!!field.prefix?.onClick\"\r\n                    >{{ field.prefix!.text }}</span\r\n                  >\r\n                }\r\n              </div>\r\n            }\r\n            <input\r\n              [type]=\"field.type === 'number' && field.isCurrency ? 'text' : 'number'\"\r\n              [name]=\"field.name\"\r\n              [id]=\"uniqueId\"\r\n              [required]=\"field.required || false\"\r\n              [placeholder]=\"field.placeholder || '' | translate\"\r\n              [disabled]=\"isDisabled || false\"\r\n              [readOnly]=\"field.readOnly || false\"\r\n              [formControl]=\"control\"\r\n              [min]=\"field.type === 'number' && !field.isCurrency ? 0 : null\"\r\n              [max]=\"field.type === 'number' && !field.isCurrency ? field.max : null\"\r\n              [step]=\"field.type === 'number' && !field.isCurrency ? field.step || 1 : null\"\r\n              (input)=\"handleNumberInput($event, field.minFractionDigits || 2)\"\r\n              (paste)=\"handleNumberPaste($event, field.minFractionDigits || 2)\"\r\n              (blur)=\"handleBlur($event, field.minFractionDigits || 2)\"\r\n              (keydown)=\"handleNumberKeydown($event, field.isCurrency || false)\"\r\n              class=\"fn-input-field\"\r\n              [ngClass]=\"{\r\n                disabled: control.disabled || field.readOnly,\r\n                error: control.touched && control.errors,\r\n                success: field['hasSuccessBorder'] && control.valid && !control.errors,\r\n                'pl-10': field.prefix?.icon && !field.prefix?.text,\r\n                'has-prefix-text': !!field.prefix?.text,\r\n                'pr-10': field.suffix?.icon && !field.suffix?.text,\r\n                'has-suffix-text': !!field.suffix?.text,\r\n                'text-[24px] font-bold': field.isCurrency,\r\n              }\"\r\n              [ngStyle]=\"{\r\n                color: field.valueColor || 'inherit',\r\n                'font-size': field.valueSize || '',\r\n              }\"\r\n              (focus)=\"onFocus()\"\r\n            />\r\n          </div>\r\n        }\r\n\r\n        <!-- Default Input Field (text, email, etc.) -->\r\n        @default {\r\n          <div class=\"relative z-0\">\r\n            @if (field.prefix?.icon || field.prefix?.text) {\r\n              <div\r\n                class=\"absolute left-0 top-1/2 -translate-y-1/2 pl-3 flex items-center gap-2\"\r\n                style=\"z-index: 1\"\r\n              >\r\n                @if (field.prefix?.icon) {\r\n                  <fn-icon\r\n                    [name]=\"field.prefix?.icon || ''\"\r\n                    [variant]=\"'Line'\"\r\n                    [size]=\"'medium'\"\r\n                    [color]=\"control.disabled ? '#c4cdd2' : '#03182b'\"\r\n                  ></fn-icon>\r\n                }\r\n                @if (field.prefix?.text) {\r\n                  <span\r\n                    class=\"text-sm font-medium whitespace-nowrap\"\r\n                    [ngStyle]=\"{ color: control.disabled ? '#c4cdd2' : '#03182b' }\"\r\n                    (click)=\"prefixClick($event)\"\r\n                    (keydown.enter)=\"prefixClick($event)\"\r\n                    [attr.tabindex]=\"field.prefix?.onClick ? 0 : null\"\r\n                    [class.cursor-pointer]=\"!!field.prefix?.onClick\"\r\n                    >{{ field.prefix!.text }}</span\r\n                  >\r\n                }\r\n              </div>\r\n            }\r\n            <input\r\n              [type]=\"field.type\"\r\n              [name]=\"field.name\"\r\n              [id]=\"uniqueId\"\r\n              [required]=\"field.required || false\"\r\n              [placeholder]=\"field.placeholder || '' | translate\"\r\n              [disabled]=\"field.disabled || false\"\r\n              [readOnly]=\"field.readOnly || false\"\r\n              [formControl]=\"control\"\r\n              (input)=\"\r\n                isEmailField\r\n                  ? handleEmailInput($event)\r\n                  : field.type === 'text' && isAlphanumeric\r\n                    ? handleAlphanumericInput($event)\r\n                    : handleInput($event)\r\n              \"\r\n              (blur)=\"handleBlur($event)\"\r\n              class=\"fn-input-field\"\r\n              [ngClass]=\"{\r\n                disabled: control.disabled || field.readOnly,\r\n                error: control.touched && control.errors,\r\n                success: field['hasSuccessBorder'] && control.valid && !control.errors,\r\n                'pl-10': field.prefix?.icon && !field.prefix?.text,\r\n                'has-prefix-text': !!field.prefix?.text,\r\n                'pr-10':\r\n                  (field.icon || field.suffix?.icon) && !field.isCopyText && !field.suffix?.text,\r\n                'has-suffix-text': !!field.suffix?.text,\r\n                'pr-20':\r\n                  field.isCopyText && !field.hasGenerateKey && !(field.icon || field.suffix?.icon),\r\n                'pr-30':\r\n                  field.isCopyText && field.hasGenerateKey && !(field.icon || field.suffix?.icon),\r\n                'pr-32':\r\n                  field.isCopyText && field.hasGenerateKey && (field.icon || field.suffix?.icon),\r\n              }\"\r\n              [ngStyle]=\"{\r\n                color: field.valueColor || 'inherit',\r\n                'font-size': field.valueSize || '',\r\n              }\"\r\n              (focus)=\"onFocus()\"\r\n            />\r\n\r\n            <div class=\"absolute right-0 top-1/2 -translate-y-1/2 flex items-center pr-3 gap-2\">\r\n              @if (field.isCopyText) {\r\n                <span\r\n                  class=\"cursor-pointer\"\r\n                  (click)=\"copyText($event)\"\r\n                  (keydown)=\"copyText($event)\"\r\n                >\r\n                  <fn-icon [name]=\"'two-square'\" [color]=\"'#03182b'\"></fn-icon>\r\n                </span>\r\n                @if (field.hasGenerateKey) {\r\n                  <span\r\n                    class=\"cursor-pointer\"\r\n                    (click)=\"field.onGenerateKey && field.onGenerateKey()\"\r\n                    (keydown)=\"field.onGenerateKey && field.onGenerateKey()\"\r\n                  >\r\n                    <fn-icon [name]=\"'round-arrow-top-left'\" [color]=\"'#03182b'\"></fn-icon>\r\n                  </span>\r\n                }\r\n              }\r\n\r\n              @if (field.icon || field.suffix?.icon || field.suffix?.text) {\r\n                @if (field.suffix?.text) {\r\n                  <span\r\n                    class=\"text-sm font-medium whitespace-nowrap\"\r\n                    [ngStyle]=\"{ color: control.disabled ? '#c4cdd2' : '#03182b' }\"\r\n                    (click)=\"suffixClick($event)\"\r\n                    (keydown.enter)=\"suffixClick($event)\"\r\n                    [attr.tabindex]=\"field.suffix?.onClick ? 0 : null\"\r\n                    [class.cursor-pointer]=\"!!field.suffix?.onClick\"\r\n                    >{{ field.suffix!.text }}</span\r\n                  >\r\n                }\r\n                @if (field.icon || field.suffix?.icon) {\r\n                  <fn-icon\r\n                    [name]=\"field.suffix?.icon || field.icon?.name || ''\"\r\n                    [variant]=\"field.icon?.variant || 'Line'\"\r\n                    [size]=\"field.icon ? getIconSizeName(field.icon.size) : 'medium'\"\r\n                    [color]=\"control.disabled ? '#c4cdd2' : '#03182b'\"\r\n                  ></fn-icon>\r\n                }\r\n              }\r\n            </div>\r\n          </div>\r\n        }\r\n      }\r\n    </ng-container>\r\n\r\n    @if (showFormFieldMessage(control, helperText)) {\r\n      <fn-field-message\r\n        [field]=\"field\"\r\n        [control]=\"control\"\r\n        [hasFocus]=\"hasFocus()\"\r\n      ></fn-field-message>\r\n    }\r\n  </div>\r\n\r\n  <!-- Integrated Toast Notifications -->\r\n  <fn-toast></fn-toast>\r\n}\r\n", styles: [".fn-input-container{display:flex;flex-direction:column;width:100%}.fn-label{font-family:Outfit,sans-serif;font-weight:500;margin-bottom:8px;display:flex;align-items:center}.fn-label span{color:#c4cdd2;font-size:14px;font-weight:400}.fn-input-field{width:100%;height:48px;padding:0 16px;background-color:#f7f8f9;border-bottom:1.5px solid #c4cdd2;font-family:Outfit,sans-serif;font-size:16px;color:#03182b;transition:all .2s ease;border-radius:4px 4px 0 0}.fn-input-field::placeholder{color:#c4cdd2}.fn-input-field:focus{outline:none;background-color:#fff;border-bottom-color:#03182b}.fn-input-field.error{border-bottom-color:#ef3e42}.fn-input-field.success{border-bottom-color:#6cc24a}.fn-input-field.disabled{background-color:#eef0f2;color:#c4cdd2;cursor:not-allowed}.fn-input-field.pl-10{padding-left:40px!important}.fn-input-field.has-prefix-text{padding-left:100px!important}.fn-input-field.pr-10{padding-right:40px!important}.fn-input-field.has-suffix-text{padding-right:120px!important}.fn-input-field.pr-12{padding-right:44px!important}.fn-input-field.pr-20{padding-right:72px!important}.fn-input-field.pr-24{padding-right:88px!important}.fn-input-field.pr-30{padding-right:104px!important}.fn-input-field.pr-32{padding-right:120px!important}.icon-container{display:inline-flex;align-items:center;justify-content:center;width:24px;height:24px}.relative{position:relative}.absolute{position:absolute}.right-0{right:0}.left-0{left:0}.top-1\\/2{top:50%}.-translate-y-1\\/2{transform:translateY(-50%)}.flex{display:flex}.items-center{align-items:center}.justify-between{justify-content:space-between}.gap-1{gap:.25rem}.gap-2{gap:.5rem}.pr-3{padding-right:.75rem}.pl-3{padding-left:.75rem}.mt-2{margin-top:.5rem}.space-y-1>:not([hidden])~:not([hidden]){margin-top:.25rem}.text-xs{font-size:.75rem}.font-medium{font-weight:500}.h-1\\.5{height:.375rem}.w-full{width:100%}.bg-\\[\\#eef0f2\\]{background-color:#eef0f2}.rounded-full{border-radius:9999px}.overflow-hidden{overflow:hidden}.transition-all{transition-property:all}.duration-300{transition-duration:.3s}.cursor-pointer{cursor:pointer}.opacity-50{opacity:.5}.cursor-not-allowed{cursor:not-allowed}.text-ellipsis{text-overflow:ellipsis}.whitespace-nowrap{white-space:nowrap}\n"], dependencies: [{ kind: "ngmodule", type: FormsModule }, { kind: "directive", type: i1.DefaultValueAccessor, selector: "input:not([type=checkbox])[formControlName],textarea[formControlName],input:not([type=checkbox])[formControl],textarea[formControl],input:not([type=checkbox])[ngModel],textarea[ngModel],[ngDefaultControl]" }, { kind: "directive", type: i1.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { kind: "directive", type: i1.RequiredValidator, selector: ":not([type=checkbox])[required][formControlName],:not([type=checkbox])[required][formControl],:not([type=checkbox])[required][ngModel]", inputs: ["required"] }, { kind: "ngmodule", type: CommonModule }, { kind: "directive", type: i2.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "directive", type: i2.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }, { kind: "ngmodule", type: ReactiveFormsModule }, { kind: "directive", type: i1.FormControlDirective, selector: "[formControl]", inputs: ["formControl", "disabled", "ngModel"], outputs: ["ngModelChange"], exportAs: ["ngForm"] }, { kind: "ngmodule", type: TranslateModule }, { kind: "pipe", type: i3.TranslatePipe, name: "translate" }, { kind: "component", type: FNLabel, selector: "fn-label", inputs: ["field", "label", "required", "className", "class", "for", "color", "variant", "statusLabel", "ngClass", "hideOptional"] }, { kind: "component", type: FNFieldMessage, selector: "fn-field-message", inputs: ["field", "control", "hasFocus"] }, { kind: "component", type: FNIconComponent, selector: "fn-icon", inputs: ["name", "variant", "size", "color", "strokeWidth", "disabled", "className", "basePath"] }, { kind: "component", type: FNToast, selector: "fn-toast" }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.19", ngImport: i0, type: FNInput, decorators: [{
            type: Component,
            args: [{ selector: 'fn-input', standalone: true, imports: [
                        FormsModule,
                        CommonModule,
                        ReactiveFormsModule,
                        TranslateModule,
                        FNLabel,
                        FNFieldMessage,
                        FNIconComponent,
                        FNToast,
                    ], changeDetection: ChangeDetectionStrategy.OnPush, template: "@if (field.name && isVisible() && !field.hidden) {\r\n  <div class=\"fn-input-container\">\r\n    @if (field.type! !== 'hidden') {\r\n      <fn-label\r\n        [for]=\"uniqueId\"\r\n        [label]=\"field.label\"\r\n        [required]=\"field.required || false\"\r\n        [hideOptional]=\"field.hideOptional || false\"\r\n        [color]=\"field.color || '#03182b'\"\r\n        [variant]=\"field.labelVariant || 'p1'\"\r\n        [statusLabel]=\"field.statusLabel\"\r\n      ></fn-label>\r\n    }\r\n\r\n    <ng-container>\r\n      @switch (field.type) {\r\n        <!-- Textarea Field -->\r\n        @case ('textarea') {\r\n          <textarea\r\n            #fnTextarea\r\n            style=\"resize: none\"\r\n            [name]=\"field.name\"\r\n            [id]=\"uniqueId\"\r\n            [required]=\"field.required || false\"\r\n            [placeholder]=\"field.placeholder || '' | translate\"\r\n            [disabled]=\"field.disabled || false\"\r\n            [rows]=\"field.rows || 1\"\r\n            [readonly]=\"field.readOnly || false\"\r\n            [formControl]=\"control\"\r\n            (input)=\"handleTextArea($event)\"\r\n            (blur)=\"handleBlur($event)\"\r\n            class=\"fn-input-field\"\r\n            [ngClass]=\"{\r\n              disabled: control.disabled || field.readOnly,\r\n              error: control.touched && control.errors,\r\n              success: field['hasSuccessBorder'] && control.valid && !control.errors,\r\n            }\"\r\n            [ngStyle]=\"{\r\n              color: field.valueColor || 'inherit',\r\n              'font-size': field.valueSize || '',\r\n            }\"\r\n            (focus)=\"onFocus()\"\r\n          ></textarea>\r\n        }\r\n\r\n        <!-- Password Field -->\r\n        @case ('password') {\r\n          <div class=\"relative w-full\">\r\n            <div\r\n              class=\"absolute left-0 top-1/2 -translate-y-1/2 pl-3 flex items-center gap-2\"\r\n              style=\"z-index: 1\"\r\n            >\r\n              @if (field.prefix?.icon) {\r\n                <fn-icon\r\n                  [name]=\"field.prefix?.icon || ''\"\r\n                  [variant]=\"'Line'\"\r\n                  [size]=\"'medium'\"\r\n                  [color]=\"control.disabled ? '#c4cdd2' : '#03182b'\"\r\n                ></fn-icon>\r\n              }\r\n              @if (field.prefix?.text) {\r\n                <span\r\n                  class=\"text-sm font-medium whitespace-nowrap\"\r\n                  [ngStyle]=\"{ color: control.disabled ? '#c4cdd2' : '#03182b' }\"\r\n                  (click)=\"prefixClick($event)\"\r\n                  (keydown.enter)=\"prefixClick($event)\"\r\n                  [attr.tabindex]=\"field.prefix?.onClick ? 0 : null\"\r\n                  [class.cursor-pointer]=\"!!field.prefix?.onClick\"\r\n                  >{{ field.prefix!.text }}</span\r\n                >\r\n              }\r\n            </div>\r\n            <input\r\n              [type]=\"showPassword ? 'text' : 'password'\"\r\n              [id]=\"uniqueId\"\r\n              [required]=\"field.required || false\"\r\n              [disabled]=\"field.disabled || false\"\r\n              [formControl]=\"control\"\r\n              [placeholder]=\"field.placeholder || '' | translate\"\r\n              [readonly]=\"field.readOnly || false\"\r\n              (input)=\"isAlphanumeric ? handleAlphanumericInput($event) : handleInput($event)\"\r\n              (focus)=\"onPasswordFocus()\"\r\n              (blur)=\"handlePasswordBlur($event)\"\r\n              class=\"fn-input-field overflow-hidden text-ellipsis\"\r\n              [ngClass]=\"{\r\n                disabled: control.disabled || field.readOnly,\r\n                error: control.touched && control.errors,\r\n                success: field['hasSuccessBorder'] && control.valid,\r\n                'pl-10': field.prefix?.icon && !field.prefix?.text,\r\n                'has-prefix-text': !!field.prefix?.text,\r\n                'pr-12':\r\n                  (field.toggleMask && !field.isCopyText) ||\r\n                  (!field.toggleMask && field.isCopyText),\r\n                'pr-24': field.toggleMask && field.isCopyText && !field.hasGenerateKey,\r\n                'pr-32': field.toggleMask && field.isCopyText && field.hasGenerateKey,\r\n              }\"\r\n              [ngStyle]=\"{\r\n                color: field.valueColor || 'inherit',\r\n                'font-size': field.valueSize || '',\r\n              }\"\r\n            />\r\n\r\n            <div class=\"absolute right-0 top-1/2 -translate-y-1/2 flex items-center pr-3 gap-2\">\r\n              @if (field.toggleMask && !field.isCopyText) {\r\n                <span\r\n                  class=\"cursor-pointer\"\r\n                  (click)=\"togglePasswordVisibility()\"\r\n                  (keydown)=\"togglePasswordVisibility()\"\r\n                >\r\n                  <fn-icon\r\n                    [name]=\"showPassword ? 'eye-open' : 'eye-close'\"\r\n                    [variant]=\"'Line'\"\r\n                    [size]=\"'large'\"\r\n                    [color]=\"'#03182b'\"\r\n                  ></fn-icon>\r\n                </span>\r\n              } @else if (field.isCopyText && !field.toggleMask) {\r\n                <span\r\n                  class=\"cursor-pointer\"\r\n                  (click)=\"copyText($event)\"\r\n                  (keydown)=\"copyText($event)\"\r\n                >\r\n                  <fn-icon [name]=\"'two-square'\" [color]=\"'#03182b'\"></fn-icon>\r\n                </span>\r\n                @if (field.hasGenerateKey) {\r\n                  <span\r\n                    class=\"cursor-pointer\"\r\n                    (click)=\"field.onGenerateKey && field.onGenerateKey()\"\r\n                    (keydown)=\"field.onGenerateKey && field.onGenerateKey()\"\r\n                  >\r\n                    <fn-icon\r\n                      [name]=\"'round-arrow-top-left'\"\r\n                      [variant]=\"'Line'\"\r\n                      [size]=\"'large'\"\r\n                      [color]=\"'#03182b'\"\r\n                    ></fn-icon>\r\n                  </span>\r\n                }\r\n              } @else if (field.toggleMask && field.isCopyText) {\r\n                <span\r\n                  class=\"cursor-pointer\"\r\n                  (click)=\"togglePasswordVisibility()\"\r\n                  (keydown)=\"togglePasswordVisibility()\"\r\n                >\r\n                  <fn-icon\r\n                    [name]=\"showPassword ? 'eye-open' : 'eye-close'\"\r\n                    [variant]=\"'Line'\"\r\n                    [size]=\"'large'\"\r\n                    [color]=\"'#03182b'\"\r\n                  ></fn-icon>\r\n                </span>\r\n                <span\r\n                  class=\"cursor-pointer\"\r\n                  (click)=\"copyText($event)\"\r\n                  (keydown)=\"copyText($event)\"\r\n                >\r\n                  <fn-icon [name]=\"'two-square'\" [color]=\"'#03182b'\"></fn-icon>\r\n                </span>\r\n                @if (field.hasGenerateKey) {\r\n                  <span\r\n                    class=\"cursor-pointer\"\r\n                    (click)=\"field.onGenerateKey && field.onGenerateKey()\"\r\n                    (keydown)=\"field.onGenerateKey && field.onGenerateKey()\"\r\n                    [ngClass]=\"{ 'opacity-50 cursor-not-allowed': !control.value }\"\r\n                  >\r\n                    <fn-icon [name]=\"'round-arrow-top-left'\" [color]=\"'#03182b'\"></fn-icon>\r\n                  </span>\r\n                }\r\n              }\r\n            </div>\r\n          </div>\r\n\r\n          @if (field.feedback && isPasswordFocused) {\r\n            <div class=\"mt-2 space-y-1\">\r\n              <div class=\"flex items-center justify-between text-xs\">\r\n                <span class=\"font-medium\" [ngClass]=\"strengthClass\">\r\n                  {{ passwordStrengthLabel | translate }}\r\n                </span>\r\n                <div class=\"flex items-center gap-1\">\r\n                  <fn-icon\r\n                    [name]=\"passwordStrengthPercent > 50 ? 'check-circle' : 'info-circle'\"\r\n                    [variant]=\"'Line'\"\r\n                    [size]=\"'extrasmall'\"\r\n                    [color]=\"\r\n                      passwordStrengthPercent <= 25\r\n                        ? '#ef3e42'\r\n                        : passwordStrengthPercent <= 75\r\n                          ? '#ff9f00'\r\n                          : '#6cc24a'\r\n                    \"\r\n                  ></fn-icon>\r\n                </div>\r\n              </div>\r\n              <div class=\"h-1.5 w-full bg-[#eef0f2] rounded-full overflow-hidden\">\r\n                <div\r\n                  class=\"h-full transition-all duration-300 rounded-full\"\r\n                  [ngClass]=\"{\r\n                    'bg-[#ef3e42]': passwordStrengthPercent <= 25,\r\n                    'bg-[#ff9f00]': passwordStrengthPercent > 25 && passwordStrengthPercent <= 75,\r\n                    'bg-[#6cc24a]': passwordStrengthPercent > 75,\r\n                  }\"\r\n                  [style.width.%]=\"passwordStrengthPercent\"\r\n                ></div>\r\n              </div>\r\n            </div>\r\n          }\r\n        }\r\n\r\n        <!-- Number Field -->\r\n        @case ('number') {\r\n          <div class=\"relative w-full\">\r\n            @if (field.prefix?.icon || field.prefix?.text) {\r\n              <div\r\n                class=\"absolute left-0 top-1/2 -translate-y-1/2 pl-3 flex items-center gap-2\"\r\n                style=\"z-index: 1\"\r\n              >\r\n                @if (field.prefix?.icon) {\r\n                  <fn-icon\r\n                    [name]=\"field.prefix?.icon || ''\"\r\n                    [variant]=\"'Line'\"\r\n                    [size]=\"'medium'\"\r\n                    [color]=\"control.disabled ? '#c4cdd2' : '#03182b'\"\r\n                  ></fn-icon>\r\n                }\r\n                @if (field.prefix?.text) {\r\n                  <span\r\n                    class=\"text-sm font-medium whitespace-nowrap\"\r\n                    [ngStyle]=\"{ color: control.disabled ? '#c4cdd2' : '#03182b' }\"\r\n                    (click)=\"prefixClick($event)\"\r\n                    (keydown.enter)=\"prefixClick($event)\"\r\n                    [attr.tabindex]=\"field.prefix?.onClick ? 0 : null\"\r\n                    [class.cursor-pointer]=\"!!field.prefix?.onClick\"\r\n                    >{{ field.prefix!.text }}</span\r\n                  >\r\n                }\r\n              </div>\r\n            }\r\n            <input\r\n              [type]=\"field.type === 'number' && field.isCurrency ? 'text' : 'number'\"\r\n              [name]=\"field.name\"\r\n              [id]=\"uniqueId\"\r\n              [required]=\"field.required || false\"\r\n              [placeholder]=\"field.placeholder || '' | translate\"\r\n              [disabled]=\"isDisabled || false\"\r\n              [readOnly]=\"field.readOnly || false\"\r\n              [formControl]=\"control\"\r\n              [min]=\"field.type === 'number' && !field.isCurrency ? 0 : null\"\r\n              [max]=\"field.type === 'number' && !field.isCurrency ? field.max : null\"\r\n              [step]=\"field.type === 'number' && !field.isCurrency ? field.step || 1 : null\"\r\n              (input)=\"handleNumberInput($event, field.minFractionDigits || 2)\"\r\n              (paste)=\"handleNumberPaste($event, field.minFractionDigits || 2)\"\r\n              (blur)=\"handleBlur($event, field.minFractionDigits || 2)\"\r\n              (keydown)=\"handleNumberKeydown($event, field.isCurrency || false)\"\r\n              class=\"fn-input-field\"\r\n              [ngClass]=\"{\r\n                disabled: control.disabled || field.readOnly,\r\n                error: control.touched && control.errors,\r\n                success: field['hasSuccessBorder'] && control.valid && !control.errors,\r\n                'pl-10': field.prefix?.icon && !field.prefix?.text,\r\n                'has-prefix-text': !!field.prefix?.text,\r\n                'pr-10': field.suffix?.icon && !field.suffix?.text,\r\n                'has-suffix-text': !!field.suffix?.text,\r\n                'text-[24px] font-bold': field.isCurrency,\r\n              }\"\r\n              [ngStyle]=\"{\r\n                color: field.valueColor || 'inherit',\r\n                'font-size': field.valueSize || '',\r\n              }\"\r\n              (focus)=\"onFocus()\"\r\n            />\r\n          </div>\r\n        }\r\n\r\n        <!-- Default Input Field (text, email, etc.) -->\r\n        @default {\r\n          <div class=\"relative z-0\">\r\n            @if (field.prefix?.icon || field.prefix?.text) {\r\n              <div\r\n                class=\"absolute left-0 top-1/2 -translate-y-1/2 pl-3 flex items-center gap-2\"\r\n                style=\"z-index: 1\"\r\n              >\r\n                @if (field.prefix?.icon) {\r\n                  <fn-icon\r\n                    [name]=\"field.prefix?.icon || ''\"\r\n                    [variant]=\"'Line'\"\r\n                    [size]=\"'medium'\"\r\n                    [color]=\"control.disabled ? '#c4cdd2' : '#03182b'\"\r\n                  ></fn-icon>\r\n                }\r\n                @if (field.prefix?.text) {\r\n                  <span\r\n                    class=\"text-sm font-medium whitespace-nowrap\"\r\n                    [ngStyle]=\"{ color: control.disabled ? '#c4cdd2' : '#03182b' }\"\r\n                    (click)=\"prefixClick($event)\"\r\n                    (keydown.enter)=\"prefixClick($event)\"\r\n                    [attr.tabindex]=\"field.prefix?.onClick ? 0 : null\"\r\n                    [class.cursor-pointer]=\"!!field.prefix?.onClick\"\r\n                    >{{ field.prefix!.text }}</span\r\n                  >\r\n                }\r\n              </div>\r\n            }\r\n            <input\r\n              [type]=\"field.type\"\r\n              [name]=\"field.name\"\r\n              [id]=\"uniqueId\"\r\n              [required]=\"field.required || false\"\r\n              [placeholder]=\"field.placeholder || '' | translate\"\r\n              [disabled]=\"field.disabled || false\"\r\n              [readOnly]=\"field.readOnly || false\"\r\n              [formControl]=\"control\"\r\n              (input)=\"\r\n                isEmailField\r\n                  ? handleEmailInput($event)\r\n                  : field.type === 'text' && isAlphanumeric\r\n                    ? handleAlphanumericInput($event)\r\n                    : handleInput($event)\r\n              \"\r\n              (blur)=\"handleBlur($event)\"\r\n              class=\"fn-input-field\"\r\n              [ngClass]=\"{\r\n                disabled: control.disabled || field.readOnly,\r\n                error: control.touched && control.errors,\r\n                success: field['hasSuccessBorder'] && control.valid && !control.errors,\r\n                'pl-10': field.prefix?.icon && !field.prefix?.text,\r\n                'has-prefix-text': !!field.prefix?.text,\r\n                'pr-10':\r\n                  (field.icon || field.suffix?.icon) && !field.isCopyText && !field.suffix?.text,\r\n                'has-suffix-text': !!field.suffix?.text,\r\n                'pr-20':\r\n                  field.isCopyText && !field.hasGenerateKey && !(field.icon || field.suffix?.icon),\r\n                'pr-30':\r\n                  field.isCopyText && field.hasGenerateKey && !(field.icon || field.suffix?.icon),\r\n                'pr-32':\r\n                  field.isCopyText && field.hasGenerateKey && (field.icon || field.suffix?.icon),\r\n              }\"\r\n              [ngStyle]=\"{\r\n                color: field.valueColor || 'inherit',\r\n                'font-size': field.valueSize || '',\r\n              }\"\r\n              (focus)=\"onFocus()\"\r\n            />\r\n\r\n            <div class=\"absolute right-0 top-1/2 -translate-y-1/2 flex items-center pr-3 gap-2\">\r\n              @if (field.isCopyText) {\r\n                <span\r\n                  class=\"cursor-pointer\"\r\n                  (click)=\"copyText($event)\"\r\n                  (keydown)=\"copyText($event)\"\r\n                >\r\n                  <fn-icon [name]=\"'two-square'\" [color]=\"'#03182b'\"></fn-icon>\r\n                </span>\r\n                @if (field.hasGenerateKey) {\r\n                  <span\r\n                    class=\"cursor-pointer\"\r\n                    (click)=\"field.onGenerateKey && field.onGenerateKey()\"\r\n                    (keydown)=\"field.onGenerateKey && field.onGenerateKey()\"\r\n                  >\r\n                    <fn-icon [name]=\"'round-arrow-top-left'\" [color]=\"'#03182b'\"></fn-icon>\r\n                  </span>\r\n                }\r\n              }\r\n\r\n              @if (field.icon || field.suffix?.icon || field.suffix?.text) {\r\n                @if (field.suffix?.text) {\r\n                  <span\r\n                    class=\"text-sm font-medium whitespace-nowrap\"\r\n                    [ngStyle]=\"{ color: control.disabled ? '#c4cdd2' : '#03182b' }\"\r\n                    (click)=\"suffixClick($event)\"\r\n                    (keydown.enter)=\"suffixClick($event)\"\r\n                    [attr.tabindex]=\"field.suffix?.onClick ? 0 : null\"\r\n                    [class.cursor-pointer]=\"!!field.suffix?.onClick\"\r\n                    >{{ field.suffix!.text }}</span\r\n                  >\r\n                }\r\n                @if (field.icon || field.suffix?.icon) {\r\n                  <fn-icon\r\n                    [name]=\"field.suffix?.icon || field.icon?.name || ''\"\r\n                    [variant]=\"field.icon?.variant || 'Line'\"\r\n                    [size]=\"field.icon ? getIconSizeName(field.icon.size) : 'medium'\"\r\n                    [color]=\"control.disabled ? '#c4cdd2' : '#03182b'\"\r\n                  ></fn-icon>\r\n                }\r\n              }\r\n            </div>\r\n          </div>\r\n        }\r\n      }\r\n    </ng-container>\r\n\r\n    @if (showFormFieldMessage(control, helperText)) {\r\n      <fn-field-message\r\n        [field]=\"field\"\r\n        [control]=\"control\"\r\n        [hasFocus]=\"hasFocus()\"\r\n      ></fn-field-message>\r\n    }\r\n  </div>\r\n\r\n  <!-- Integrated Toast Notifications -->\r\n  <fn-toast></fn-toast>\r\n}\r\n", styles: [".fn-input-container{display:flex;flex-direction:column;width:100%}.fn-label{font-family:Outfit,sans-serif;font-weight:500;margin-bottom:8px;display:flex;align-items:center}.fn-label span{color:#c4cdd2;font-size:14px;font-weight:400}.fn-input-field{width:100%;height:48px;padding:0 16px;background-color:#f7f8f9;border-bottom:1.5px solid #c4cdd2;font-family:Outfit,sans-serif;font-size:16px;color:#03182b;transition:all .2s ease;border-radius:4px 4px 0 0}.fn-input-field::placeholder{color:#c4cdd2}.fn-input-field:focus{outline:none;background-color:#fff;border-bottom-color:#03182b}.fn-input-field.error{border-bottom-color:#ef3e42}.fn-input-field.success{border-bottom-color:#6cc24a}.fn-input-field.disabled{background-color:#eef0f2;color:#c4cdd2;cursor:not-allowed}.fn-input-field.pl-10{padding-left:40px!important}.fn-input-field.has-prefix-text{padding-left:100px!important}.fn-input-field.pr-10{padding-right:40px!important}.fn-input-field.has-suffix-text{padding-right:120px!important}.fn-input-field.pr-12{padding-right:44px!important}.fn-input-field.pr-20{padding-right:72px!important}.fn-input-field.pr-24{padding-right:88px!important}.fn-input-field.pr-30{padding-right:104px!important}.fn-input-field.pr-32{padding-right:120px!important}.icon-container{display:inline-flex;align-items:center;justify-content:center;width:24px;height:24px}.relative{position:relative}.absolute{position:absolute}.right-0{right:0}.left-0{left:0}.top-1\\/2{top:50%}.-translate-y-1\\/2{transform:translateY(-50%)}.flex{display:flex}.items-center{align-items:center}.justify-between{justify-content:space-between}.gap-1{gap:.25rem}.gap-2{gap:.5rem}.pr-3{padding-right:.75rem}.pl-3{padding-left:.75rem}.mt-2{margin-top:.5rem}.space-y-1>:not([hidden])~:not([hidden]){margin-top:.25rem}.text-xs{font-size:.75rem}.font-medium{font-weight:500}.h-1\\.5{height:.375rem}.w-full{width:100%}.bg-\\[\\#eef0f2\\]{background-color:#eef0f2}.rounded-full{border-radius:9999px}.overflow-hidden{overflow:hidden}.transition-all{transition-property:all}.duration-300{transition-duration:.3s}.cursor-pointer{cursor:pointer}.opacity-50{opacity:.5}.cursor-not-allowed{cursor:not-allowed}.text-ellipsis{text-overflow:ellipsis}.whitespace-nowrap{white-space:nowrap}\n"] }]
        }], ctorParameters: () => [{ type: i0.ChangeDetectorRef }], propDecorators: { field: [{
                type: Input
            }], helperHandle: [{
                type: Input
            }], toastService: [{
                type: Input
            }], currencyMeta: [{
                type: Input
            }], defaultLocale: [{
                type: Input
            }], form: [{
                type: Input
            }], valueChange: [{
                type: Output
            }], fieldBlur: [{
                type: Output
            }], textareaElement: [{
                type: ViewChild,
                args: ['fnTextarea']
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { DEFAULT_CURRENCY_META, FNInput, FN_TOAST_SERVICE };
