// fn-label/fesm2022/fn-label.mjs
import * as i2 from '@angular/common';
import { CommonModule } from '@angular/common';
import * as i0 from '@angular/core';
import { Input, Component } from '@angular/core';
import * as i1 from '@ngx-translate/core';
import { TranslateModule } from '@ngx-translate/core';

class FNLabel {
    translate;
    // 🧩 Inputs
    field;
    label = '';
    required = true;
    className = '';
    class = '';
    for = '';
    color = ''; // accepts hex, rgb, var(--token), or Tailwind class
    variant = 'p1';
    statusLabel;
    ngClass = {};
    hideOptional = false;
    constructor(translate) {
        this.translate = translate;
    }
    ngOnInit() {
        this.initialize();
        if (this.label !== undefined && this.label !== null && this.label !== '') {
            if (Array.isArray(this.label)) {
                this.label = this.label.toString().split(',').join(', ');
            }
            else if (typeof this.label === 'string' || typeof this.label === 'number') {
                this.label = this.label.toString();
            }
            else {
                this.label = '';
            }
        }
    }
    ngOnChanges(changes) {
        if (changes['required'] || changes['label']) {
            this.initialize();
        }
    }
    getEndSymbol(label) {
        return label.toString().includes('required') && label.toString().includes('FNFieldMessage')
            ? '.'
            : '';
    }
    getTranslatedLabel(label) {
        if (label === undefined || label === null || label === '')
            return ''; // handle undefined/null/empty upfront
        try {
            const labelStr = label.toString();
            return `${labelStr
                .split('::')
                .map((part) => part.trim())
                .filter((part) => !!part) // 🔹 skip empty parts
                .map((part) => this.translate.instant(part))
                .join(' ')}${this.getEndSymbol(labelStr)}`;
        }
        catch (e) {
            console.error('Error translating label:', label, e);
            return '';
        }
    }
    initialize() {
        // reserved for initialization logic if needed later
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.2", ngImport: i0, type: FNLabel, deps: [{ token: i1.TranslateService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "21.2.2", type: FNLabel, isStandalone: true, selector: "fn-label", inputs: { field: "field", label: "label", required: "required", className: "className", class: "class", for: "for", color: "color", variant: "variant", statusLabel: "statusLabel", ngClass: "ngClass", hideOptional: "hideOptional" }, usesOnChanges: true, ngImport: i0, template: `
    @if (
      (label !== undefined && label !== null && label !== '') ||
      (field?.label !== undefined && field?.label !== null && field?.label !== '')
    ) {
      <label
        [for]="for || label"
        [ngClass]="ngClass"
        [class]="
          'fn-label !whitespace-normal ' +
          (statusLabel || field?.statusLabel
            ? statusLabel || field?.statusLabel
            : field?.variant || variant || '') +
          ' ' +
          (class || className || field?.className || '')
        "
        [style.color]="color || null"
      >
        {{ getTranslatedLabel(label || field?.label || '') }}
        @if (!required && !hideOptional && !field?.hideOptional) {
          <span>&nbsp;(Optional)</span>
        }
      </label>
    }
  `, isInline: true, styles: [".fn-label{color:var(--Base-100);display:flex;letter-spacing:0}.fn-label.h1Bold{font-size:24px;font-weight:700;line-height:32px}.fn-label.h1{font-size:24px;font-weight:400;line-height:32px}.fn-label.h2Bold{font-size:20px;line-height:24px;font-weight:700}.fn-label.h2{font-size:20px;line-height:24px;font-weight:400}.fn-label.h3Bold{font-size:18px;line-height:24px;font-weight:700}.fn-label.h3,.fn-label.p1{font-size:18px;line-height:24px;font-weight:400}.fn-label.p1Bold{font-size:18px;line-height:24px;font-weight:700}.fn-label.p2{font-size:16px;font-weight:400;line-height:24px}.fn-label.p2Bold{font-size:16px;font-weight:700;line-height:24px}.fn-label.p3{font-size:14px;line-height:20px;font-weight:400}.fn-label.p3Bold{font-size:14px;line-height:20px;font-weight:700}.fn-label.p4{font-size:12px;line-height:16px;font-weight:400}.fn-label.p4Bold{font-size:12px;line-height:16px;font-weight:700}.fn-label.Standard{height:24px!important;border-radius:4px!important;padding:3px 8px 5px!important;border:1px solid #78c5ff!important;color:#0067b1!important;background:#def1ff!important;font-weight:700!important;font-size:12px!important;line-height:16px!important;width:max-content}.fn-label.Warning{height:24px!important;border-radius:4px!important;padding:3px 8px 5px!important;border:1px solid #f7cfa8!important;color:#c0410b!important;background:#fdebda!important;font-weight:700!important;font-size:12px!important;line-height:16px!important;width:max-content}.fn-label.Success{height:24px!important;border-radius:4px!important;padding:3px 8px 5px!important;border:1px solid #b5e1a4!important;color:#065c00!important;background:#e2f3db!important;font-weight:700!important;font-size:12px!important;line-height:16px!important;width:max-content}.fn-label.Hot{height:24px!important;border-radius:4px!important;padding:3px 8px 5px!important;border:1px solid #fc91a3!important;color:#bf1f21!important;background:#fed7db!important;font-weight:700!important;font-size:12px!important;line-height:16px!important;width:max-content}.fn-label.Disabled{height:24px!important;border-radius:4px!important;padding:3px 8px 5px!important;border:1px solid #c4cdd2!important;color:#4f5d6b!important;background:#eef0f2!important;font-weight:700!important;font-size:12px!important;line-height:16px!important;width:max-content}.fn-label.RHBPremierBanking{height:24px!important;border-radius:4px!important;padding:3px 8px 5px!important;border:1px solid #e0e4e9!important;color:#c4cdd2!important;background:#002353!important;font-weight:700!important;font-size:12px!important;line-height:16px!important;width:max-content}.fn-label.interactionLarge{font-size:16px;line-height:24px;font-weight:400}.fn-label.interactionLargeBold{font-size:16px;line-height:24px;font-weight:700}.fn-label.interactionMedium{font-size:14px;line-height:20px;font-weight:400}.fn-label.interactionMediumBold{font-size:14px;line-height:20px;font-weight:700}.fn-label.interactionSmall{font-size:12px;line-height:16px;font-weight:400}.fn-label.interactionSmallBold{font-size:12px;line-height:16px;font-weight:700}.fn-label.avatarLabel{font-family:Lato,sans-serif;font-weight:400;font-style:normal;font-size:28.8px;line-height:38.4px;letter-spacing:0%;text-align:center}\n"], dependencies: [{ kind: "ngmodule", type: CommonModule }, { kind: "directive", type: i2.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "ngmodule", type: TranslateModule }] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.2", ngImport: i0, type: FNLabel, decorators: [{
            type: Component,
            args: [{ selector: 'fn-label', standalone: true, imports: [CommonModule, TranslateModule], template: `
    @if (
      (label !== undefined && label !== null && label !== '') ||
      (field?.label !== undefined && field?.label !== null && field?.label !== '')
    ) {
      <label
        [for]="for || label"
        [ngClass]="ngClass"
        [class]="
          'fn-label !whitespace-normal ' +
          (statusLabel || field?.statusLabel
            ? statusLabel || field?.statusLabel
            : field?.variant || variant || '') +
          ' ' +
          (class || className || field?.className || '')
        "
        [style.color]="color || null"
      >
        {{ getTranslatedLabel(label || field?.label || '') }}
        @if (!required && !hideOptional && !field?.hideOptional) {
          <span>&nbsp;(Optional)</span>
        }
      </label>
    }
  `, styles: [".fn-label{color:var(--Base-100);display:flex;letter-spacing:0}.fn-label.h1Bold{font-size:24px;font-weight:700;line-height:32px}.fn-label.h1{font-size:24px;font-weight:400;line-height:32px}.fn-label.h2Bold{font-size:20px;line-height:24px;font-weight:700}.fn-label.h2{font-size:20px;line-height:24px;font-weight:400}.fn-label.h3Bold{font-size:18px;line-height:24px;font-weight:700}.fn-label.h3,.fn-label.p1{font-size:18px;line-height:24px;font-weight:400}.fn-label.p1Bold{font-size:18px;line-height:24px;font-weight:700}.fn-label.p2{font-size:16px;font-weight:400;line-height:24px}.fn-label.p2Bold{font-size:16px;font-weight:700;line-height:24px}.fn-label.p3{font-size:14px;line-height:20px;font-weight:400}.fn-label.p3Bold{font-size:14px;line-height:20px;font-weight:700}.fn-label.p4{font-size:12px;line-height:16px;font-weight:400}.fn-label.p4Bold{font-size:12px;line-height:16px;font-weight:700}.fn-label.Standard{height:24px!important;border-radius:4px!important;padding:3px 8px 5px!important;border:1px solid #78c5ff!important;color:#0067b1!important;background:#def1ff!important;font-weight:700!important;font-size:12px!important;line-height:16px!important;width:max-content}.fn-label.Warning{height:24px!important;border-radius:4px!important;padding:3px 8px 5px!important;border:1px solid #f7cfa8!important;color:#c0410b!important;background:#fdebda!important;font-weight:700!important;font-size:12px!important;line-height:16px!important;width:max-content}.fn-label.Success{height:24px!important;border-radius:4px!important;padding:3px 8px 5px!important;border:1px solid #b5e1a4!important;color:#065c00!important;background:#e2f3db!important;font-weight:700!important;font-size:12px!important;line-height:16px!important;width:max-content}.fn-label.Hot{height:24px!important;border-radius:4px!important;padding:3px 8px 5px!important;border:1px solid #fc91a3!important;color:#bf1f21!important;background:#fed7db!important;font-weight:700!important;font-size:12px!important;line-height:16px!important;width:max-content}.fn-label.Disabled{height:24px!important;border-radius:4px!important;padding:3px 8px 5px!important;border:1px solid #c4cdd2!important;color:#4f5d6b!important;background:#eef0f2!important;font-weight:700!important;font-size:12px!important;line-height:16px!important;width:max-content}.fn-label.RHBPremierBanking{height:24px!important;border-radius:4px!important;padding:3px 8px 5px!important;border:1px solid #e0e4e9!important;color:#c4cdd2!important;background:#002353!important;font-weight:700!important;font-size:12px!important;line-height:16px!important;width:max-content}.fn-label.interactionLarge{font-size:16px;line-height:24px;font-weight:400}.fn-label.interactionLargeBold{font-size:16px;line-height:24px;font-weight:700}.fn-label.interactionMedium{font-size:14px;line-height:20px;font-weight:400}.fn-label.interactionMediumBold{font-size:14px;line-height:20px;font-weight:700}.fn-label.interactionSmall{font-size:12px;line-height:16px;font-weight:400}.fn-label.interactionSmallBold{font-size:12px;line-height:16px;font-weight:700}.fn-label.avatarLabel{font-family:Lato,sans-serif;font-weight:400;font-style:normal;font-size:28.8px;line-height:38.4px;letter-spacing:0%;text-align:center}\n"] }]
        }], ctorParameters: () => [{ type: i1.TranslateService }], propDecorators: { field: [{
                type: Input
            }], label: [{
                type: Input
            }], required: [{
                type: Input
            }], className: [{
                type: Input
            }], class: [{
                type: Input
            }], for: [{
                type: Input
            }], color: [{
                type: Input
            }], variant: [{
                type: Input
            }], statusLabel: [{
                type: Input
            }], ngClass: [{
                type: Input
            }], hideOptional: [{
                type: Input
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { FNLabel };
