// fn-toast/fesm2022/fn-toast.mjs
import { CommonModule } from '@angular/common';
import * as i0 from '@angular/core';
import { Injectable, signal, inject, Component } from '@angular/core';
import * as i1 from '@ngx-translate/core';
import { TranslateModule } from '@ngx-translate/core';
import { FNIconComponent } from '@arpudhabotupload/fn-icon-angular';
import { BehaviorSubject, Subscription } from 'rxjs';

class ToastService {
    toastsSubject = new BehaviorSubject([]);
    toasts$ = this.toastsSubject.asObservable();
    generateId() {
        return crypto.randomUUID();
    }
    addToast(toast) {
        const newToast = {
            ...toast,
            id: this.generateId(),
            isVisible: false,
        };
        const currentToasts = this.toastsSubject.value;
        this.toastsSubject.next([...currentToasts, newToast]);
        // Trigger animation
        setTimeout(() => {
            const toasts = this.toastsSubject.value;
            const toastIndex = toasts.findIndex((t) => t.id === newToast.id);
            if (toastIndex !== -1) {
                toasts[toastIndex].isVisible = true;
                this.toastsSubject.next([...toasts]);
            }
        }, 10);
        // Auto remove if duration is specified
        if (toast.duration && toast.duration > 0) {
            setTimeout(() => {
                this.remove(newToast.id);
            }, toast.duration);
        }
        return newToast.id;
    }
    success(message, title, duration = 3000) {
        return this.addToast({
            type: 'success',
            title,
            message,
            duration,
        });
    }
    error(message, title, duration = 5000) {
        return this.addToast({
            type: 'error',
            title,
            message,
            duration,
        });
    }
    warn(message, title, duration = 5000) {
        return this.addToast({
            type: 'warn',
            title,
            message,
            duration,
        });
    }
    info(message, title, duration = 5000) {
        return this.addToast({
            type: 'info',
            title,
            message,
            duration,
        });
    }
    remove(id) {
        const currentToasts = this.toastsSubject.value;
        const filteredToasts = currentToasts.filter((toast) => toast.id !== id);
        this.toastsSubject.next(filteredToasts);
    }
    clear() {
        this.toastsSubject.next([]);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.19", ngImport: i0, type: ToastService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
    static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "19.2.19", ngImport: i0, type: ToastService, providedIn: 'root' });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.19", ngImport: i0, type: ToastService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
class FNToast {
    toasts = signal([]);
    subscription = new Subscription();
    toastService = inject(ToastService);
    constructor() { }
    ngOnInit() {
        this.subscription.add(this.toastService.toasts$.subscribe((toasts) => {
            this.toasts.set(toasts);
        }));
    }
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
    removeToast(id) {
        this.toastService.remove(id);
    }
    getIconName(toast) {
        switch (toast.type) {
            case 'success':
                return 'check-circle-colour';
            case 'warn':
                return 'info-circle-colour';
            case 'error':
                return 'exclamation-circle-colour';
            case 'info':
                return 'info-circle-colour';
            default:
                return 'info-circle-colour';
        }
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.19", ngImport: i0, type: FNToast, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "19.2.19", type: FNToast, isStandalone: true, selector: "fn-toast", ngImport: i0, template: "<div class=\"toast-container\">\r\n  @for (toast of toasts(); track toast.id) {\r\n    <div class=\"toast toast-{{ toast.type }}\" [class.toast-enter]=\"toast.isVisible\">\r\n      <span class=\"toast-icon\">\r\n        <fn-icon [name]=\"getIconName(toast)\" variant=\"Colour\" size=\"medium\"></fn-icon>\r\n      </span>\r\n      <div class=\"toast-content\">\r\n        <span class=\"toast-message\">\r\n          {{ toast.message | translate }}\r\n        </span>\r\n      </div>\r\n      <button class=\"toast-close\" (click)=\"removeToast(toast.id)\" aria-label=\"Close\">\r\n        <fn-icon name=\"cross\" variant=\"Line\" size=\"medium\" color=\"#ffffff\"></fn-icon>\r\n      </button>\r\n    </div>\r\n  }\r\n</div>\r\n", styles: [".toast-container{position:fixed;top:24px;left:50%;transform:translate(-50%);z-index:9999;width:100%;max-width:336px;display:flex;flex-direction:column;gap:8px;pointer-events:none}.toast{display:flex;align-items:center;background:#1c2f40;border-radius:8px;padding:12px;width:336px;max-width:100%;opacity:0;transform:translateY(-20px);transition:all .3s cubic-bezier(.4,0,.2,1);pointer-events:auto}.toast.toast-enter{opacity:1;transform:translateY(0)}.toast-icon{width:20px;height:20px;border-radius:50%;display:flex;align-items:center;justify-content:center;margin-right:8px;flex-shrink:0;background:transparent}.toast-icon fn-icon{width:20px;height:20px;display:block}.toast-content{flex:1;min-width:0}.toast-message{font-size:14px;font-weight:500;line-height:20px;color:#fff;display:block}.toast-close{background:none;border:none;font-size:22px;color:#fff;opacity:.7;cursor:pointer;padding:0;margin-left:8px;width:24px;height:24px;display:flex;align-items:center;justify-content:center;flex-shrink:0;transition:opacity .2s;pointer-events:auto}.toast-close:hover{opacity:1}.toast.toast-leave{opacity:0;transform:translateY(-20px)}\n"], dependencies: [{ kind: "ngmodule", type: CommonModule }, { kind: "component", type: FNIconComponent, selector: "fn-icon", inputs: ["name", "variant", "size", "color", "strokeWidth", "disabled", "className", "basePath"] }, { kind: "ngmodule", type: TranslateModule }, { kind: "pipe", type: i1.TranslatePipe, name: "translate" }] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.19", ngImport: i0, type: FNToast, decorators: [{
            type: Component,
            args: [{ selector: 'fn-toast', standalone: true, imports: [CommonModule, FNIconComponent, TranslateModule], template: "<div class=\"toast-container\">\r\n  @for (toast of toasts(); track toast.id) {\r\n    <div class=\"toast toast-{{ toast.type }}\" [class.toast-enter]=\"toast.isVisible\">\r\n      <span class=\"toast-icon\">\r\n        <fn-icon [name]=\"getIconName(toast)\" variant=\"Colour\" size=\"medium\"></fn-icon>\r\n      </span>\r\n      <div class=\"toast-content\">\r\n        <span class=\"toast-message\">\r\n          {{ toast.message | translate }}\r\n        </span>\r\n      </div>\r\n      <button class=\"toast-close\" (click)=\"removeToast(toast.id)\" aria-label=\"Close\">\r\n        <fn-icon name=\"cross\" variant=\"Line\" size=\"medium\" color=\"#ffffff\"></fn-icon>\r\n      </button>\r\n    </div>\r\n  }\r\n</div>\r\n", styles: [".toast-container{position:fixed;top:24px;left:50%;transform:translate(-50%);z-index:9999;width:100%;max-width:336px;display:flex;flex-direction:column;gap:8px;pointer-events:none}.toast{display:flex;align-items:center;background:#1c2f40;border-radius:8px;padding:12px;width:336px;max-width:100%;opacity:0;transform:translateY(-20px);transition:all .3s cubic-bezier(.4,0,.2,1);pointer-events:auto}.toast.toast-enter{opacity:1;transform:translateY(0)}.toast-icon{width:20px;height:20px;border-radius:50%;display:flex;align-items:center;justify-content:center;margin-right:8px;flex-shrink:0;background:transparent}.toast-icon fn-icon{width:20px;height:20px;display:block}.toast-content{flex:1;min-width:0}.toast-message{font-size:14px;font-weight:500;line-height:20px;color:#fff;display:block}.toast-close{background:none;border:none;font-size:22px;color:#fff;opacity:.7;cursor:pointer;padding:0;margin-left:8px;width:24px;height:24px;display:flex;align-items:center;justify-content:center;flex-shrink:0;transition:opacity .2s;pointer-events:auto}.toast-close:hover{opacity:1}.toast.toast-leave{opacity:0;transform:translateY(-20px)}\n"] }]
        }], ctorParameters: () => [] });

/**
 * Generated bundle index. Do not edit.
 */

export { FNToast, ToastService };
